{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analysis of lysine deserts in selected Orthologous Groups (OGs) from the eggNOG5 database"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Import libraries and read data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "pd.options.display.max_columns = None\n",
    "import numpy as np\n",
    "import math\n",
    "import statistics\n",
    "import scipy.stats as stats\n",
    "\n",
    "import pickle\n",
    "import gzip\n",
    "import os\n",
    "import re\n",
    "\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import collections\n",
    "from IPython.display import Markdown, display\n",
    "\n",
    "from Bio import AlignIO\n",
    "from ete3 import NCBITaxa, Tree, TreeStyle, AttrFace, TextFace, add_face_to_node\n",
    "\n",
    "ncbi = NCBITaxa()\n",
    "\n",
    "######### Read preprocessed data obtained by running Download_data.ipynb #########\n",
    "\n",
    "# Read dictionary of lists of selected MSA files for each taxonomic group\n",
    "with open(\"results/additional/filtered_msa_dict.pickle\", \"rb\") as handle: \n",
    "    selected_MSAs = pickle.load(handle)\n",
    "\n",
    "# Read dictionary mapping each MSA file to its Gene Symbol\n",
    "with open(\"results/additional/SequenceID_to_GeneSymbol.pickle\", \"rb\") as handle: \n",
    "    seqID_gene_symbol = pickle.load(handle)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Search for lysine deserts in OGs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.1. Set parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "eukaryota = ['4893', '6236', '7214', '9989', '9604']\n",
    "\n",
    "######################################### LYSINE-LESS REGION ########################################\n",
    "# Define lysine desert region minimum length; integer or float\n",
    "desert_min_lengths = [150, 0.5]\n",
    "# Define minimum fraction of characters in MSA column which are not Lysine\n",
    "min_desert_fraction = 0.6\n",
    "# Define minimum fraction of characters in MSA column which are not gaps\n",
    "min_nongaps_fraction = 0.6\n",
    "# Define maximum length of continous MSA region for which gap fractions are below min_nongaps_fraction\n",
    "insertion_length = 15\n",
    "# Define minimum fraction of MSA sequences which have completely lysine devoid region of given length\n",
    "min_fraction_seq_lysine_less_region = 0.6 \n",
    "\n",
    "########################################### LYSINE CLUSTER ##########################################\n",
    "# Define minimum fraction of total sequence lysines to be in close proximity to form lysine cluster\n",
    "lysine_cluster_fraction = 0.8\n",
    "# Define maximum fraction of sequence length to contain defined above fraction of all lysines\n",
    "lysine_cluster_seq_fraction = 0.2\n",
    "# Define fraction at the beggining of sequence to be considered as N-terminal\n",
    "N_term_end_fraction = 0.2\n",
    "# Define fraction at the end of sequence to be considered as C-terminal\n",
    "C_term_end_fraction = 0.8\n",
    "# Minimum percentage of sequences in MSA file to consider MSA as lysine cluster containing\n",
    "# A less strict threshold for analyzing multiple eukaryotic groups (vectors)\n",
    "LYSINE_CLUSTER_MSA_THRESHOLD_VECTORS = 0.6 \n",
    "\n",
    "############################################ CANONICAL AA ###########################################\n",
    "CANON_AA = ['A', 'G', 'P', 'V', 'L', 'I', 'M', 'C', 'F', 'Y', \n",
    "            'W', 'H', 'K', 'R', 'Q', 'N', 'E', 'D', 'S', 'T']\n",
    "\n",
    "######################################## REFERENCE ORGANISMS ########################################\n",
    "REFERENCE_ORGANISMS = {'4893': '4932', '6236': '6239', '7214': '7227', '9989': '10090', '9604':'9606'} "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.2. Main code\n",
    "\n",
    "\n",
    "### Lysine desert search\n",
    "\n",
    "We define lysine deserts as either:\n",
    "1. lysine-devoid region of minimum 150 aa length\n",
    "2. lysine-devoid region of minimum 50% median MSA sequence length\n",
    "\n",
    "The algorithm distinguishes two types of MSA containing lysine desert region:\n",
    "* Type **Aligned**, where the lysine desert region of defined length is well conserved in part of the alignment\n",
    "* Type **Anywhere**, where the lysine desert region of defined length is not conserved in part of the alignment, but most of sequences contain lysine desert region (in any part of the sequence)\n",
    "\n",
    "The abovementioned lysine desert region classification was introduced due to varying quality of alignments, where some contain very long gaps regions.\n",
    "\n",
    "The algorithm checks in the first place if MSA has lysine desert region type Aligned, if not checks for lysine desert region type Anywhere.\n",
    "\n",
    "> **_NOTE:_** Both lysine deserts types (Aligned/Anywhere) are subsequently used to plot phylogenetics trees.\n",
    "\n",
    "#### Default author's parameters\n",
    "\n",
    "* Type **Aligned**\n",
    "  * MSA contains continous region of declared minimum length in which each column has minimum 70% of non-lysines and non-gaps. Small continous insertions (meaning that a few columns in a row contain majority of gaps) are allowed. At least 60% of sequences in that region has to be completely lysine-devoid.\n",
    "* Type **Anywhere**\n",
    "  * Each sequence of MSA is checked for presence of completely lysine-less chunk of declared minimum length (it may be anywhere in sequence). At least 80% of sequences have to contain such chunks.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#### Algorithm implementation\n",
    "\n",
    "The algorithm is as follows:\n",
    "\n",
    "1. Take MSA, create two masks:\n",
    "\n",
    "    * Mask 1: if in MSA column at least 60% of characters were not K (lysine) assign 1, 0 otherwise\n",
    "    * Mask 2: if in MSA column at least 60% of characters were not gaps assign 1, 0 otherwise\n",
    "\n",
    "\n",
    "2. Merge both masks. \n",
    "\n",
    "    * Final Mask = Mask 1 with the exception, that if Mask 2 had fragments with more than declared (default: 15) continous \"0\" values, then Final Mask will also had \"0\" values in that fragments. \n",
    "\n",
    "\n",
    "3. In the Final Mask, filter fragments with continous \"1\" values of given length - these will be our lysine desert regions type *Aligned*.\n",
    "\n",
    "Please note that in type *Aligned*, there may be a small number of Lysines in lysine desert aligned region. However defined fraction of sequences in aligned region (default: at least 60%) will be completely lysine-devoid.\n",
    "\n",
    "\n",
    "4. If no single lysine desert region type *Aligned* was found, look for type *Anywhere*. \n",
    "\n",
    "    In order to do so, take each sequence from MSA and search for lysine-less chunk of given length (in any part of the sequence). If more than defined fraction of sequences possess such chunks, we classify the MSA as type *Anywhere*.\n",
    "\n",
    "\n",
    "---\n",
    "\n",
    "### Lysine cluster search\n",
    "\n",
    "In addition to lysine desert region search, each sequence (**if it contains at least 2 lysines**) is checked for the presence of lysine cluster.\n",
    "\n",
    "By default lysine cluster is defined as ⌊80% * total number of lysines in sequence⌋ within 20% of sequence length. \n",
    "\n",
    "We distinguish three types of lysine cluster, according to its position in sequence:\n",
    "* anywhere in sequence\n",
    "* at N-end, defined as first 20% of sequence\n",
    "* at C-end, defined as last 20% of sequence\n",
    "\n",
    "For each MSA, fraction of occcurrence of lysine cluster of defined type among all MSA sequences is calculated."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "class FindDesertMSA():\n",
    "\n",
    "    def __init__(self, MSA):\n",
    "        \"\"\"Open, read and parse alignment file\n",
    "        \n",
    "        :param MSA: path to alignment file\n",
    "        :type MSA: str\n",
    "        :return: parsed alignment file object\n",
    "        :rtype: Bio.Align.MultipleSeqAlignment\n",
    "        \"\"\"\n",
    "        \n",
    "        f = gzip.open(MSA, \"rt\")\n",
    "        alignment = AlignIO.read(f, \"fasta\")\n",
    "        f.close()\n",
    "            \n",
    "        self.MSA = MSA\n",
    "        self.og = MSA.split('/')[-2]\n",
    "        self.msa_id = MSA.split('/')[-1].split('.')[0]\n",
    "        self.matrix = np.array([list(seq) for seq in alignment], order=\"F\")\n",
    "        self.names = np.array([alignment[name].id for name in range(len(alignment))])\n",
    "        self.length = len(self.matrix) # No of sequences in alignment\n",
    "        \n",
    "    def _get_seq(self, seq_id, matrix=None):\n",
    "        \"\"\"Get sequence without dashes from alignment\n",
    "        \n",
    "        :param seq_id: sequence index in alignment matrix\n",
    "        :type seq_id: int\n",
    "        :param matrix: matrix\n",
    "        :type matrix: np.array\n",
    "        :return: sequence\n",
    "        :rtype: str\n",
    "        \"\"\"\n",
    "        \n",
    "        if matrix is None:\n",
    "            matrix = self.matrix\n",
    "\n",
    "        return ''.join(matrix[seq_id]).replace('-', '')\n",
    "        \n",
    "    def _calculate_desert_fracs(self):\n",
    "        \"\"\"Calculate for each column of MSA frequencies of non-K aa (frequency of all aa except K, including gaps)\"\"\"\n",
    "        \n",
    "        desert_fracs = (np.count_nonzero(self.matrix != 'K', axis = 0) / self.length) \n",
    "        return desert_fracs\n",
    "\n",
    "    def _calculate_nongap_fracs(self):\n",
    "        \"\"\"Calculate for each column of MSA frequencies of non-gaps characters.\"\"\"\n",
    "        \n",
    "        nongap_fracs = (np.count_nonzero(self.matrix != '-', axis = 0) / self.length)\n",
    "        \n",
    "        return nongap_fracs\n",
    "    \n",
    "    def _calculate_MSA_taxons(self):\n",
    "        \"\"\"Calculate number of MSA taxons.\"\"\"\n",
    "        \n",
    "        MSA_taxon_set = set()\n",
    "        for taxon in self.names:\n",
    "            taxonID = taxon.split('.')[0]\n",
    "            MSA_taxon_set.add(taxonID)\n",
    "\n",
    "        return len(MSA_taxon_set)\n",
    "    \n",
    "    def calculate_median_seq_length(self, matrix=None):\n",
    "        \"\"\"Calculate median of sequence length of a given matrix\n",
    "        \n",
    "        :param matrix: matrix\n",
    "        :type matrix: np.array\n",
    "        :return: median of sequence length of a given matrix\n",
    "        :rtype: float\n",
    "        \"\"\"\n",
    "        \n",
    "        if matrix is None:\n",
    "            matrix = self.matrix\n",
    "            \n",
    "        all_seq_lengths = []\n",
    "        \n",
    "        for seq_index in range(len(matrix)):\n",
    "            all_seq_lengths.append(len(self._get_seq(seq_index, matrix))) \n",
    "        \n",
    "        return statistics.median(all_seq_lengths)\n",
    "    \n",
    "    def get_gene_names(self, ref_org=False):\n",
    "        \"\"\"Get Gene Symbols for sequences in MSA, or alternatively for reference organism only\n",
    "        \n",
    "        :return: gene names list\n",
    "        :rtype: list\n",
    "        \"\"\"\n",
    "        \n",
    "        gene_names = set()\n",
    "        \n",
    "        for seq_name in self.names:\n",
    "            if seq_name in seqID_gene_symbol.keys():\n",
    "                \n",
    "                if ref_org:\n",
    "                    if seq_name.split('.')[0] == ref_org:\n",
    "                        gene_names.add(seqID_gene_symbol[seq_name])\n",
    "                else:\n",
    "                    gene_names.add(seqID_gene_symbol[seq_name])\n",
    "        \n",
    "        return list(gene_names)\n",
    "    \n",
    "    def get_COG(self):\n",
    "        \"\"\"Get MSA COG annotation\n",
    "\n",
    "        :return: COG annotations letter(s)\n",
    "        :rtype: list\n",
    "        \"\"\"\n",
    "\n",
    "        df_cog = pd.read_csv('data/{}/{}_annotations.tsv.gz'.format(self.og, self.og), sep = '\\t', header = None)\n",
    "        cog = df_cog[df_cog[1] == self.msa_id][2].values\n",
    "        return ''.join(cog)\n",
    "\n",
    "    def _check_frac_of_seqs_with_desert(self, matrix):\n",
    "        \"\"\"Calculate fraction of sequences in MSA trimmed to lysine desert region which have absolutely no lysine\n",
    "        \n",
    "        :param matrix: lysine desert region, trimmed matrix\n",
    "        :type matrix: np.array\n",
    "        :return: fraction of sequences which in the given trimmed MSA have abolutely no lysine\n",
    "        :rtype: list\n",
    "        \"\"\"\n",
    "        \n",
    "        seq_with_complete_desert = 0\n",
    "        \n",
    "        for seq_index in range(len(matrix)):\n",
    "            seq = self._get_seq(seq_index, matrix)\n",
    "            K_number = seq.count('K')\n",
    "            \n",
    "            if K_number == 0:\n",
    "                seq_with_complete_desert += 1\n",
    "\n",
    "        return round(seq_with_complete_desert / self.length, 2)\n",
    "        \n",
    "    def _score_desert(self, start, end, desert_nongaps_fracs):\n",
    "        \"\"\"Score lysine desert region\n",
    "        \n",
    "        :param start: lysine desert region start index in MSA\n",
    "        :type start: int\n",
    "        :param end: lysine desert region end index in MSA\n",
    "        :type end: int\n",
    "        :param desert_nongaps_fracs: nongap_fracs * desert_fracs\n",
    "        :type desert_nongaps_fracs: np.array\n",
    "        :return: average lysine desert conservation, average lysine desert length\n",
    "        :rtype: list\n",
    "        \"\"\"\n",
    "        \n",
    "        idx = [i for i in range(start, end)]\n",
    "        trimmed_matrix = self.matrix[:, idx]\n",
    " \n",
    "        median_desert_length = self.calculate_median_seq_length(trimmed_matrix)\n",
    "        average_desert_conservation = np.sum(desert_nongaps_fracs[start:end]) / len(trimmed_matrix[0,])\n",
    "        \n",
    "        return round(average_desert_conservation, 2), round(median_desert_length, 2)\n",
    "\n",
    "    \n",
    "    def _check_MSA_for_potential_desert(self, min_length):\n",
    "        \"\"\"If no lysine desert region type Aligned was found, try to find it considering only MSA sequences.\n",
    "        In other words, check if each sequence of MSA has anywhere in its sequence a defined lysine desert region. \n",
    "        If more than declared fraction of sequences (default: 0.8) contain such region, MSA is considerd as lysine desert.\n",
    "        \n",
    "        :rtype min_length: int or float\n",
    "        :return: fraction of sequences that contain lysine desert region of declared length, median of longest lysine desert regions lengths\n",
    "        :rtype: float, float\n",
    "        \"\"\"\n",
    "\n",
    "        seq_list_no_gaps = []\n",
    "        \n",
    "        for seq_index in range(len(self.matrix)):\n",
    "            seq_list_no_gaps.append(self._get_seq(seq_index, self.matrix)) \n",
    "            \n",
    "        potential_desert_counter = 0\n",
    "        desert_lengths = []\n",
    "        \n",
    "        for seq in range(len(seq_list_no_gaps)):\n",
    "            longest_desert = 0\n",
    "            seq_filtered = ''.join(['1' if aa != 'K' else '0' for aa in seq_list_no_gaps[seq]])\n",
    "            desert_indices_all = [(des.start(), des.end()) for des in re.finditer('1+', seq_filtered)] # get lysine-less sequence chunks\n",
    "\n",
    "            for indice in desert_indices_all:\n",
    "\n",
    "                start, end = indice\n",
    "                length = int(end) - int(start)\n",
    "                if length >= min_length:\n",
    "                    if length > longest_desert:\n",
    "                        longest_desert = length\n",
    "            \n",
    "            if longest_desert != 0:\n",
    "                potential_desert_counter += 1\n",
    "                desert_lengths.append(longest_desert)\n",
    "\n",
    "        if potential_desert_counter == 0:\n",
    "            return 0, 0\n",
    "        else:\n",
    "            return round(potential_desert_counter/self.length, 2), statistics.median(desert_lengths)\n",
    "        \n",
    "    def calc_median_lysine_less_region_percentage(self):\n",
    "        \"\"\"Calculate median percentage of longest lysine-less region per sequence in MSA\n",
    "        \n",
    "        :return: median percentage of longest lysine-less region per sequence in MSA\n",
    "        :rtype: float\n",
    "        \"\"\"\n",
    "        \n",
    "        lysine_less_lengths = []\n",
    "        seq_list_no_gaps = []\n",
    "        \n",
    "        for seq_index in range(len(self.matrix)):\n",
    "            seq_list_no_gaps.append(self._get_seq(seq_index, self.matrix)) \n",
    "        \n",
    "        for seq in range(len(seq_list_no_gaps)):\n",
    "            longest_desert = 0\n",
    "            seq_filtered = ''.join(['1' if aa != 'K' else '0' for aa in seq_list_no_gaps[seq]])\n",
    "            desert_indices = [(match.start(), match.end()) for match in re.finditer('1+', seq_filtered)]\n",
    "            \n",
    "            for indice in desert_indices:\n",
    "\n",
    "                start, end = indice\n",
    "                length = int(end) - int(start)\n",
    "                if length >= longest_desert:\n",
    "                    longest_desert = length\n",
    "            \n",
    "            lysine_less_lengths.append(round(100*(longest_desert/len(seq_filtered)), 2))\n",
    "\n",
    "        return round(statistics.median(lysine_less_lengths), 2)\n",
    "        \n",
    "    def _check_lysine_cluster(self):\n",
    "            \"\"\"Check if lysine cluster region occurs in each MSA sequence. \n",
    "            By default, lysine cluster is defined if at least 80% of sequence lysines are within 20% sequence length.\n",
    "            Minimum number of lysines in sequence is 2 to check it for lysine cluster\n",
    "\n",
    "            :return: total lysine cluster fraction per sequence in MSA, N-term lysine cluster fraction per sequence in MSA, C-term lysine cluster fraction per sequence in MSA\n",
    "            :rtype: list\n",
    "            \"\"\"\n",
    "            \n",
    "            total_cluster_no = 0\n",
    "            N_cluster_no = 0\n",
    "            C_cluster_no = 0\n",
    "\n",
    "            for seq_index in range(len(self.matrix)):\n",
    "                biggest_cluster = 0\n",
    "                left_idx = []\n",
    "                right_idx = []\n",
    "                seq = self._get_seq(seq_index)\n",
    "                cluster_max_length = lysine_cluster_seq_fraction * len(seq)\n",
    "\n",
    "                only_lysine = ''.join(['1' if aa == 'K' else '0' for aa in seq])\n",
    "                only_lysine_indices = [match.start() for match in re.finditer('1', only_lysine)]\n",
    "\n",
    "                if len(only_lysine_indices) > 2:\n",
    "                    # by default 1-lysine_cluster_fraction is 1-0.8=0.2; 80% of lysines have to be present to form lysine cluster\n",
    "                    for i in range(math.ceil(len(only_lysine_indices)*(1-lysine_cluster_fraction))+1): \n",
    "                        for j in range(1,math.ceil(len(only_lysine_indices)*(1-lysine_cluster_fraction))+2-i):\n",
    "                            if only_lysine_indices[-j] - only_lysine_indices[i] < cluster_max_length:\n",
    "                                cluster = len(only_lysine_indices[i:-j]) + 1\n",
    "                                left_idx.append(only_lysine_indices[i])\n",
    "                                right_idx.append(only_lysine_indices[-j])\n",
    "\n",
    "                    if len(right_idx) > 0:\n",
    "\n",
    "                        for right, left in zip(right_idx, left_idx):\n",
    "                            if right < N_term_end_fraction * len(seq):\n",
    "                                N_cluster_no += 1\n",
    "                                break\n",
    "                            elif left > C_term_end_fraction * len(seq):\n",
    "                                C_cluster_no += 1\n",
    "                                break\n",
    "\n",
    "                        total_cluster_no += 1\n",
    "                        \n",
    "                elif len(only_lysine_indices) == 2:\n",
    "                    if only_lysine_indices[1] - only_lysine_indices[0] < cluster_max_length:\n",
    "                        total_cluster_no += 1\n",
    "                        if only_lysine_indices[1] < N_term_end_fraction * len(seq):\n",
    "                            N_cluster_no += 1\n",
    "                        if only_lysine_indices[1] > C_term_end_fraction * len(seq):\n",
    "                            C_cluster_no += 1\n",
    "\n",
    "            return round(total_cluster_no/len(self.matrix),2), round(N_cluster_no/len(self.matrix),2), round(C_cluster_no/len(self.matrix),2)   \n",
    "    \n",
    "    def calculate_avg_aa_per_sequence(self, aa):\n",
    "        \"\"\"Calculate average number and average sequence percentage of given amino acid per sequence in MSA \n",
    "        \n",
    "        :param aa: amino acid one letter name \n",
    "        :type aa: str\n",
    "        :return: average number and average sequence percentage of given amino acid per sequence in MSA\n",
    "        :rtype: float, float\n",
    "        \"\"\"\n",
    "        \n",
    "        aa_no_in_seq = 0\n",
    "        percentages = []\n",
    "        for seq_index in range(len(self.matrix)):\n",
    "            seq = self._get_seq(seq_index)\n",
    "            aa_no_in_seq += seq.count(aa.upper())\n",
    "            percentages.append((seq.count(aa.upper())/len(seq))*100)\n",
    "            \n",
    "        return round(aa_no_in_seq/len(self.matrix),2), round(statistics.mean(percentages), 2)\n",
    "        \n",
    "    def _check_gene_in_ref_org(self):\n",
    "        \"\"\"Checks gene symbol of reference organism\n",
    "\n",
    "        :return: information gene symbol of reference organism\n",
    "        :rtype: str\n",
    "        \"\"\"\n",
    "            \n",
    "        ref_org_genes = set()\n",
    "        \n",
    "        for seq_index in range(len(self.matrix)):\n",
    "            taxon = self.names[seq_index].split('.')[0]\n",
    "            \n",
    "            if taxon in REFERENCE_ORGANISMS.values():\n",
    "\n",
    "                    if self.names[seq_index] in seqID_gene_symbol.keys():\n",
    "                        ref_org_genes.add(seqID_gene_symbol[self.names[seq_index]])\n",
    "        \n",
    "        if len(ref_org_genes) > 0:\n",
    "            ref_org_genes = list(ref_org_genes)\n",
    "            ref_org_genes = ';'.join((str(gene) for gene in ref_org_genes))\n",
    "        else:\n",
    "            ref_org_genes = None\n",
    "            \n",
    "        return ref_org_genes\n",
    "            \n",
    "    def _calculate_aa_distributions(self, matrix):\n",
    "        \"\"\"For all sequences in given matrix calculate number of each aa occurrence\n",
    "        \n",
    "        :param matrix: sequence matrix\n",
    "        :rtype matrix: np.array\n",
    "        :return: dictionary with each aa number of occurrence\n",
    "        :rtype: dict\n",
    "        \"\"\"\n",
    "        \n",
    "        aa_no = np.zeros(22) # 20 aa + unknown letter + total aa\n",
    "        aa_dict = {'A' : 0, 'G': 0, 'P': 0, 'V':0, 'L':0, 'I':0, 'M':0, 'C':0, 'F':0, 'Y':0, 'W':0, 'H':0, 'K':0, 'R':0, 'Q':0, 'N':0, 'E':0, 'D':0, 'S':0, 'T':0, 'Other':0, 'Total':0}\n",
    "        \n",
    "        for seq_index in range(self.length):\n",
    "            seq = self._get_seq(seq_index, matrix)\n",
    "            for residue in seq:\n",
    "                if residue in CANON_AA: # 20 canonical aa\n",
    "                    aa_dict[residue] += 1\n",
    "                else:\n",
    "                    aa_dict['Other'] += 1\n",
    "                aa_dict['Total'] += 1\n",
    "        \n",
    "        return aa_dict\n",
    "    \n",
    "    def find_deserts(self, desert_min_length):\n",
    "        \"\"\"Find lysine desert regions of declared length in MSA.\n",
    "        \n",
    "        :param desert_min_length: minimum lysine desert region length or sequence fraction\n",
    "        :type desert_min_length: int or float\n",
    "        :param min_desert_fraction: minimum fraction of characters in MSA column which are not lysine \n",
    "        :type min_desert_fraction: float\n",
    "        :return: list of lists for each lysine desert region in MSA (may be more than 1), containing information about its properties \n",
    "        :rtype: list\n",
    "        \"\"\"\n",
    "        \n",
    "        gene_names = self.get_gene_names()\n",
    "        nongap_fracs = self._calculate_nongap_fracs()        \n",
    "        desert_fracs = self._calculate_desert_fracs()\n",
    "        \n",
    "        # 1 if non-gap fraction in col >= min_nongaps_fraction, 0 otherwise\n",
    "        nongap_fracs_mask = ''.join(['1' if frac >= min_nongaps_fraction else '0' for frac in nongap_fracs]) \n",
    "        # 1 if non-lysine fraction in col >= min_desert_fraction, 0 otherwise\n",
    "        desert_mask = np.array(['1' if frac >= min_desert_fraction else '0' for frac in desert_fracs]) \n",
    "        # Find indices of gaps\n",
    "        gaps_indices = [(chunk.start(), chunk.end()) for chunk in re.finditer('0+', nongap_fracs_mask)] \n",
    "        long_gaps_indices = []\n",
    "        \n",
    "        for gap in gaps_indices: # save long gaps indices\n",
    "            gap_length = gap[1] - gap[0]\n",
    "            if gap_length > insertion_length:\n",
    "                indices_range = [i for i in range(gap[0], gap[1])]\n",
    "                long_gaps_indices.extend(indices_range) \n",
    "        \n",
    "        # Filter long gaps from lysine desert regions\n",
    "        np.put(desert_mask, long_gaps_indices, '0') \n",
    "        \n",
    "        desert_mask = ''.join(desert_mask)\n",
    "        # Get lysine-less regions\n",
    "        desert_indices_all = [(des.start(), des.end()) for des in re.finditer('1+', desert_mask)] \n",
    "        desert_indices_proper = []\n",
    "        \n",
    "        scores = []\n",
    "        \n",
    "        if type(desert_min_length) == float:\n",
    "            desert_min_length = int(desert_min_length * self.calculate_median_seq_length())\n",
    "\n",
    "        for des_chunk in desert_indices_all:\n",
    "            indices = [i for i in range(des_chunk[0], des_chunk[1])]\n",
    "            trimmed_matrix = self.matrix[:, indices]\n",
    " \n",
    "            median_desert_length = self.calculate_median_seq_length(trimmed_matrix)\n",
    "            \n",
    "            if median_desert_length >= desert_min_length:\n",
    "                desert_indices_proper.append((des_chunk[0], des_chunk[1]))\n",
    "        \n",
    "        if desert_indices_proper:\n",
    "            how_many_deserts_in_msa = 0\n",
    "            for idx in desert_indices_proper:\n",
    "                indices = [i for i in range(idx[0], idx[1])]\n",
    "                \n",
    "                seq_with_lysine_less_region_fraction = self._check_frac_of_seqs_with_desert(self.matrix[:, indices])\n",
    "\n",
    "                if seq_with_lysine_less_region_fraction >= min_fraction_seq_lysine_less_region:\n",
    "                    \n",
    "                    how_many_deserts_in_msa += 1\n",
    "                    desert_type = 'Aligned'\n",
    "                    ref_org_genes = self._check_gene_in_ref_org()\n",
    "                    \n",
    "                    avg_desert_conservation, median_desert_length =  self._score_desert(idx[0], idx[1], desert_fracs * nongap_fracs)\n",
    "                    \n",
    "                    # Calculate median lysine desert length / median protein length in MSA\n",
    "                    des_length_prot_length = round(median_desert_length/self.calculate_median_seq_length(),2)\n",
    "\n",
    "                    # Add MSA COG annotations\n",
    "                    desert_cog = self.get_COG()\n",
    "\n",
    "                    # Add lysine cluster information\n",
    "                    total_cluster_no, N_cluster_no, C_cluster_no = self._check_lysine_cluster()\n",
    "                    \n",
    "                    # Add avg lysine and arginine number per sequence\n",
    "                    avg_lys_no_in_seq, avg_lys_percentage_in_seq = self.calculate_avg_aa_per_sequence('K')\n",
    "                    avg_arg_no_in_seq, avg_arg_percentage_in_seq = self.calculate_avg_aa_per_sequence('R')\n",
    "                    \n",
    "                    # Calculate aa occurrence in lysine-less region\n",
    "                    aa_stats = self._calculate_aa_distributions(self.matrix[:, indices])\n",
    "                    \n",
    "                    # Renumber indices according to sequence numeration convention (from 1)\n",
    "                    start_index = int(idx[0]) + 1\n",
    "                    if int(idx[1]) == len(self.matrix[0]): # lysine desert stretches to MSA end\n",
    "                        end_index = int(idx[1])\n",
    "                    else:\n",
    "                        end_index = int(idx[1]) - 1\n",
    "    \n",
    "                    result = [','.join(gene_names), ref_org_genes, \n",
    "                              desert_type, avg_desert_conservation, int(how_many_deserts_in_msa), seq_with_lysine_less_region_fraction, \n",
    "                              self.calculate_median_seq_length(), median_desert_length, des_length_prot_length,  \n",
    "                              start_index , end_index, self.length, round(self.length / self._calculate_MSA_taxons(),2),\n",
    "                              desert_cog,\n",
    "                              total_cluster_no, N_cluster_no, C_cluster_no, \n",
    "                              avg_lys_no_in_seq, avg_lys_percentage_in_seq,\n",
    "                              avg_arg_no_in_seq, avg_arg_percentage_in_seq,\n",
    "                              int(aa_stats['G']), int(aa_stats['P']), int(aa_stats['A']), int(aa_stats['V']), int(aa_stats['L']), int(aa_stats['I']), int(aa_stats['M']),\n",
    "                              int(aa_stats['C']), int(aa_stats['F']), int(aa_stats['Y']), int(aa_stats['W']), int(aa_stats['H']), int(aa_stats['K']), int(aa_stats['R']),\n",
    "                              int(aa_stats['Q']), int(aa_stats['N']), int(aa_stats['E']), int(aa_stats['D']), int(aa_stats['S']), int(aa_stats['T']), int(aa_stats['Other']), int(aa_stats['Total'])]\n",
    "                        \n",
    "                    scores.append(result)\n",
    "           \n",
    "        if len(scores) == 0: # we did not find any Aligned lysine desert region\n",
    "\n",
    "            check_potential_desert, median_desert_length = self._check_MSA_for_potential_desert(desert_min_length)\n",
    "\n",
    "            if check_potential_desert >= min_fraction_seq_lysine_less_region:\n",
    "                \n",
    "                desert_type = 'Anywhere'\n",
    "                ref_org_genes = self._check_gene_in_ref_org()\n",
    "                \n",
    "                # Calculate median desert length / median protein length in MSA\n",
    "                des_length_prot_length = round(median_desert_length/self.calculate_median_seq_length(),2)\n",
    "\n",
    "                # Add MSA COG annotations\n",
    "                desert_cog = self.get_COG()\n",
    "\n",
    "                # Add lysine cluster information\n",
    "                total_cluster_no, N_cluster_no, C_cluster_no = self._check_lysine_cluster()\n",
    "\n",
    "                # Add avg lysine and arginine number per sequence\n",
    "                avg_lys_no_in_seq, avg_lys_percentage_in_seq = self.calculate_avg_aa_per_sequence('K')\n",
    "                avg_arg_no_in_seq, avg_arg_percentage_in_seq = self.calculate_avg_aa_per_sequence('R')\n",
    "                \n",
    "                result = [','.join(gene_names), ref_org_genes, \n",
    "                          desert_type, None, None, check_potential_desert, self.calculate_median_seq_length(), \n",
    "                          median_desert_length, des_length_prot_length, None, None, \n",
    "                          self.length, round(self.length / self._calculate_MSA_taxons(), 2),\n",
    "                          desert_cog,  \n",
    "                          total_cluster_no, N_cluster_no, C_cluster_no, \n",
    "                          avg_lys_no_in_seq, avg_lys_percentage_in_seq,\n",
    "                          avg_arg_no_in_seq, avg_arg_percentage_in_seq,\n",
    "                          None, None, None, None, None, None, None, None, None, None, None, \n",
    "                          None, None, None, None, None, None, None, None, None, None, None]\n",
    "                \n",
    "                scores.append(result)\n",
    "\n",
    "        return scores"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Test run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[['san1', 'san1', 'Anywhere', None, None, 0.6, 661, 390, 0.59, None, None, 15, 1.0, 'O', 0.93, 0.0, 0.0, 10.53, 1.6, 39.33, 5.97, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None]]\n"
     ]
    }
   ],
   "source": [
    "aln = FindDesertMSA('data/4893/4893/3S16B.raw_alg.faa.gz')\n",
    "print(aln.find_deserts(0.5))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.3. Loop through all selected OGs\n",
    "\n",
    "Perform lysine desert search on all filtered MSA files for selected eukaryotic families/order.\n",
    "\n",
    "---\n",
    "\n",
    "### Summary tables\n",
    "\n",
    "If function `calculate_lysine_deserts()` is run with argument `save_desert=True`, the summary tables for all found lysine deserts will be saved in tab-separated format (tsv).\n",
    "\n",
    "\n",
    "#### Naming convention\n",
    "\n",
    "Summary tables will be saved separately for each taxonomic rank to `results/summary_tables/group_ID`, where group_ID is eggNOG5 ID of analyzed taxonomic rank (identical to NCBI Taxonomy ID).\n",
    "\n",
    "The naming convention of summary tables is as follows: \n",
    "\n",
    "`summary_lysine_desert_150.tsv.gz` or `summary_lysine_desert_0.5.tsv.gz`, \n",
    "\n",
    "where 150 or 0.5 refer to analyzed minimum declared lysine desert length.\n",
    "\n",
    "#### Indexing\n",
    "\n",
    "Summary tables are indexed by modified MSA ID (relating to original eggNOG5 ID) as follows: \n",
    "\n",
    "`MSAId_X` \n",
    "\n",
    "where MSAId refers to MSA ID from eggNOG5 database and X indicates occurrence of MSA in summary table, e.g. if MSA contains two different lysine deserts (only in case of **Aligned** lysine desert type), it will occur twice in summary table, indexed as `MSAId_1` and `MSAId_2` for different lysine deserts regions respectively.\n",
    "\n",
    "#### Columns description\n",
    "\n",
    "Summary tables contain the following information:\n",
    "\n",
    "| Column  | Description  |\n",
    "|:--|:--|\n",
    "| **All gene symbols**  | All Gene Symbols mapped to MSA  |\n",
    "|**Reference organism gene symbols** | Gene symbols associated with reference taxon|\n",
    "| **Lysine desert type**  |  Type of found lysine desert, see Par. 3.1. of `Pipeline.ipynb` |\n",
    "|  **Avg lysine desert conservation** | Average number of columns with defined fraction of non-gaps characters (default: 70%); provides information about quantity of gaps in lysine desert region.  <br> **Only for Aligned**  lysine desert type  |\n",
    "| **Lysine deserts no**  | Number of lysine deserts regions.  <br> **Only for Aligned**  lysine desert type   |\n",
    "|**Fraction of sequences completely lysine devoid** | Fraction of sequences (either within aligned lysine desert region in Aligned type or among entire sequences in Anywhere type) which contain region completely devoid of lysine of given length|\n",
    "|**Median sequence length** | Median sequence length in MSA|\n",
    "|**Median lysine desert length** | Median sequence length of lysine desert regions| \n",
    "|**Ratio of median lysine desert to median sequence length** | Median lysine desert length divided by median sequence length|\n",
    "|**Lysine desert start index** | Start index (sequence indexed from 1) of lysine desert region.  <br> **Only for Aligned**  lysine desert type| \n",
    "|**Lysine desert end index** | End index (sequence indexed from 1) of lysine desert region.  <br> **Only for Aligned**  lysine desert type| \n",
    "|**Number of sequences** | Number of sequences in MSA|\n",
    "|**Ratio of number of sequences to number of taxons** | Number of sequences divided by number of taxons in MSA|\n",
    "|**COG** | COG annotation for MSA obtained from eggNOG5 database|\n",
    "|**Lysine cluster anywhere fraction** | Fraction of sequences in MSA which contain lysine cluster anywhere in their sequences (⌊80% of lysines⌋ located anywhere within 20% of the sequence) |\n",
    "|**Lysine cluster at N-end fraction** | Fraction of sequences in MSA which contain lysine cluster at their N-ends (⌊80% of lysines⌋ located within the first 20% of sequence)|\n",
    "|**Lysine cluster at C-end fraction** | Fraction of sequences in MSA which contain lysine cluster at their C-ends (⌊80% of lysines⌋ located within the last 20% of sequence)|\n",
    "|**Avg no of lysines** | Average number of lysines per sequence in MSA|\n",
    "|**Avg lysines fraction** | Average fraction of lysines per sequence in MSA|\n",
    "|**Avg no of arginines** | Average number of arginines per sequence in MSA|\n",
    "|**Avg arginines fraction** | Average fraction of arginines per sequence in MSA|\n",
    "|**Glycine no, Proline no, Alanine no, Valine no, Leucine no,  <br> Isoleucine no, Methionine no, Cysteine no, Phenylalanine no,  Tyrosine no, Tryptophan no, Histidine no, Lysine no, Arginine no,  <br> Glutamine no, Asparagine no, Glutamic acid no, Aspartic acid no, Serine no, Threonine no, Other no** | Counted number of given amino acid within lysine desert region.  <br>**Only for Aligned**  lysine desert type| \n",
    "|**Total no** | Total number of amino acids in lysine desert region.  <br>**Only for Aligned** lysine desert type| "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_lysine_deserts(selected_OG, desert_min_length, save_deserts=False):\n",
    "    \"\"\"Find lysine-less regions among selected MSA files\n",
    "    \n",
    "    :selected_OG: dictionary of taxonomic groups as keys and lists of filtered MSA files as values\n",
    "    :type selected_OG: dict\n",
    "    :desert_min_length: minimum lysine-less region length or sequence fraction to consider lysine desert\n",
    "    :type desert_min_length: int or float\n",
    "    :save_desert: indicates if summary table with found lysine deserts should be saved\n",
    "    :type save_desert: bool\n",
    "    :return: dictionary with list of scores for each taxonomic group\n",
    "    :rtype: dict\n",
    "    \"\"\"\n",
    "    \n",
    "    # Dictionary with number of lysine deserts containing MSA files for each taxonomic group\n",
    "    lysine_desert_total = {}\n",
    "    \n",
    "    for group in selected_OG.keys():\n",
    "        \n",
    "        group_found_deserts = {}\n",
    "        desert_counter = 0\n",
    "        \n",
    "        for MSA_id in selected_OG[group]:\n",
    "        \n",
    "            aln = FindDesertMSA('data/{}/{}/{}.raw_alg.faa.gz'.format(group, group, MSA_id))\n",
    "            scores = aln.find_deserts(desert_min_length)\n",
    "            \n",
    "            if len(scores) > 0:\n",
    "                desert_counter += 1\n",
    "            \n",
    "            for i in range(len(scores)):\n",
    "                group_found_deserts['{}_{}'.format(MSA_id, i+1)] = scores[i]\n",
    "\n",
    "        if save_deserts:\n",
    "            if len(group_found_deserts) > 0:\n",
    "                \n",
    "                save_table_path = 'results/summary_tables/{}'.format(group)\n",
    "                if not os.path.exists(save_table_path):\n",
    "                    os.makedirs(save_table_path)\n",
    "                    \n",
    "                df_csv = pd.DataFrame.from_dict(group_found_deserts, orient='index')\n",
    "                df_csv.columns = ['All gene symbols', 'Reference organism gene symbols', \n",
    "                                  'Lysine desert type', 'Avg lysine desert conservation', \n",
    "                                  'Lysine deserts no', 'Fraction of sequences completely lysine devoid ', \n",
    "                                  'Median sequence length', 'Median lysine desert length', \n",
    "                                  'Ratio of median lysine desert to median sequence length',\n",
    "                                  'Lysine desert start index', 'Lysine desert end index', \n",
    "                                  'Number of sequences', 'Ratio of number of sequences to number of taxons',\n",
    "                                  'COG',\n",
    "                                  'Lysine cluster anywhere fraction', 'Lysine cluster at N-end fraction', \n",
    "                                  'Lysine cluster at C-end fraction', \n",
    "                                  'Avg no of lysines', 'Avg lysines fraction',\n",
    "                                  'Avg no of arginines', 'Avg arginines fraction',\n",
    "                                  'Glycine no', 'Proline no', 'Alanine no', 'Valine no', 'Leucine no', 'Isoleucine no', 'Methionine no',\n",
    "                                  'Cysteine no', 'Phenylalanine no', 'Tyrosine no', 'Tryptophan no', 'Histidine no', 'Lysine no', 'Arginine no',\n",
    "                                  'Glutamine no', 'Asparagine no', 'Glutamic acid no', 'Aspartic acid no', 'Serine no', 'Threonine no',\n",
    "                                  'Other no', 'Total no']\n",
    "                df_csv.index.name = \"MSA ID\"\n",
    "                \n",
    "                df_csv.sort_values(by=['Avg lysine desert conservation'], ascending=False, inplace=True)\n",
    "                df_csv.to_csv('{}/lysine_desert_{}.tsv.gz'.format(save_table_path, desert_min_length), sep='\\t', compression='gzip')\n",
    "                \n",
    "        lysine_desert_total[group] = (len(selected_OG[group]), desert_counter)\n",
    "            \n",
    "    return lysine_desert_total"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.4. Run analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "eukaryota_dict = {}\n",
    "\n",
    "for group in selected_MSAs.keys():   \n",
    "    lineage = ncbi.get_lineage(group)\n",
    "    if 2759 in lineage:\n",
    "        eukaryota_dict[group] = selected_MSAs[group]\n",
    "        \n",
    "def run_analysis(selected_OG, pickle_out_name, save_deserts=False):\n",
    "    \n",
    "    if not os.path.exists('results/trees/pickles'):\n",
    "        os.makedirs('results/trees/pickles')\n",
    "    \n",
    "    for desert_min_length in desert_min_lengths:\n",
    "        \n",
    "        result = calculate_lysine_deserts(selected_OG, desert_min_length, save_deserts)\n",
    "        save_pickle = '{}/{}_lysine_desert_{}.pickle'.format('results/trees/pickles', pickle_out_name, desert_min_length)\n",
    "            \n",
    "        with open(save_pickle, 'wb') as handle:\n",
    "            pickle.dump(result, handle, protocol = pickle.HIGHEST_PROTOCOL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "run_analysis(eukaryota_dict, 'Eukaryota', save_deserts=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.5. Plot phylogenetic trees with annotated percentages of lysine desert "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "def draw_tree(dictionary, desert_min_length, description):\n",
    "            \n",
    "    tree = ncbi.get_topology(dictionary.keys())\n",
    "\n",
    "    def my_layout(node):\n",
    "\n",
    "        F = TextFace(node.name, fsize=22, tight_text=True)\n",
    "        add_face_to_node(F, node, column=0 , position=\"branch-right\")\n",
    "\n",
    "        if node.is_leaf():\n",
    "            sciname_face = AttrFace(\"sci_name\", fsize = 28, fgcolor = \"black\")\n",
    "            node.add_face(sciname_face, column = 0, position = \"branch-right\")\n",
    "\n",
    "            label = dictionary.get(node.name)\n",
    "\n",
    "            OG = TextFace('    ' + str(label), fsize =26, tight_text = True, bold=True)\n",
    "            add_face_to_node(OG, node, column = 0 , position = \"aligned\")\n",
    "            \n",
    "        else:\n",
    "            sciname_face = AttrFace(\"sci_name\", fsize = 24, fgcolor = \"black\")\n",
    "            node.add_face(sciname_face, column = 0, position = \"branch-right\")\n",
    "\n",
    "    ts = TreeStyle()\n",
    "    ts.layout_fn = my_layout\n",
    "    ts.show_leaf_name = False\n",
    "    ts.optimal_scale_level = \"full\"\n",
    "    tree.ladderize()\n",
    "    \n",
    "    # Save to file\n",
    "    tree.render('results/trees/{}_desert_{}.png'.format(description, desert_min_length), tree_style=ts, \n",
    "                units='px', w=3000, dpi=600)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plot phylogenetic trees for selected eukaryotic families/order and annotate them with percentages of MSA files containing lysine desert of defined length in filtered MSA files for a given taxonomic rank. \n",
    "\n",
    "> **_NOTE:_**  Only unique MSA lysine desert files are taken into account, meaning that if multiple lysine deserts were found in a single MSA file, it still will be counted as one."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "for desert_min_length in desert_min_lengths:\n",
    "    with open('results/trees/pickles/Eukaryota_lysine_desert_{}.pickle'.format(desert_min_length), 'rb') as handle:\n",
    "        Eukaryota_results = pickle.load(handle)\n",
    "\n",
    "    tree_dict = {}\n",
    "    \n",
    "    for key in Eukaryota_results.keys():\n",
    "        # Divide number of MSA with lysine desert by number of filtered MSA files\n",
    "        tree_dict[key] = str(round(100*(Eukaryota_results[key][1]/Eukaryota_results[key][0]),2))+'%' \n",
    "    \n",
    "    draw_tree(tree_dict, desert_min_length, 'Eukaryota')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. Search for evolutionary conserved lysine deserts in eukaryotic vectors\n",
    "\n",
    "We will search for lysine deserts in vectors of orthologous MSA (described in par. 5 of `Download_data.ipynb`). \n",
    "\n",
    "This approach will allow us to gain evolutionary insight into the process of lysine desert conservation on very different taxonomic levels.\n",
    "\n",
    "> **_NOTE:_** Before proceeding, run cells in par. 2.1 & 2.2.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3.1. Read data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "### Sample of constructed orthologous MSA files table for <br /> Saccharomycetaceae, Rhabditida, Drosophilidae, Rodentia, and Hominidae"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Ophistokonta OG ID</th>\n",
       "      <th>MSA paths</th>\n",
       "      <th>OG summary</th>\n",
       "      <th>Genes</th>\n",
       "      <th>Number of proteins</th>\n",
       "      <th>Number of taxons</th>\n",
       "      <th>Number of taxonomic ranks</th>\n",
       "      <th>Number of MSA files</th>\n",
       "      <th>COG</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>38B3U</td>\n",
       "      <td>9989/9989/4PSAG.raw_alg.faa.gz,9604/9604/4MZTR...</td>\n",
       "      <td>{'Rodentia': 1, 'Hominidae': 1}</td>\n",
       "      <td>usf3,kiaa2018</td>\n",
       "      <td>179</td>\n",
       "      <td>157</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>38B3V</td>\n",
       "      <td>6236/6236/40UNT.raw_alg.faa.gz,9989/9989/4PUU6...</td>\n",
       "      <td>{'Rodentia': 4, 'Rhabditida': 1, 'Hominidae': 4}</td>\n",
       "      <td>ctnnd2,arvcf,jac-1,ctnnd1,cbr-jac-1,pkp4</td>\n",
       "      <td>546</td>\n",
       "      <td>152</td>\n",
       "      <td>3</td>\n",
       "      <td>9</td>\n",
       "      <td>TW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>38B3W</td>\n",
       "      <td>9604/9604/4N2E5.raw_alg.faa.gz,9989/9989/4Q1E1...</td>\n",
       "      <td>{'Rodentia': 1, 'Hominidae': 1}</td>\n",
       "      <td>fkbp14</td>\n",
       "      <td>142</td>\n",
       "      <td>125</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>38B3X</td>\n",
       "      <td>9989/9989/4Q2TC.raw_alg.faa.gz,7214/7214/45XNB...</td>\n",
       "      <td>{'Rodentia': 3, 'Rhabditida': 1, 'Drosophilida...</td>\n",
       "      <td>lamc1,lam-2,lanb2,cbr-lam-2,dyak_lanb2,dere_la...</td>\n",
       "      <td>464</td>\n",
       "      <td>163</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>38B3Y</td>\n",
       "      <td>7214/7214/45VNG.raw_alg.faa.gz,9604/9604/4N26V...</td>\n",
       "      <td>{'Saccharomycetaceae': 1, 'Rhabditida': 1, 'Dr...</td>\n",
       "      <td>nsa2,dyak_ip259,w09c5.1,ip259</td>\n",
       "      <td>387</td>\n",
       "      <td>338</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>J</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Ophistokonta OG ID                                          MSA paths  \\\n",
       "0              38B3U  9989/9989/4PSAG.raw_alg.faa.gz,9604/9604/4MZTR...   \n",
       "1              38B3V  6236/6236/40UNT.raw_alg.faa.gz,9989/9989/4PUU6...   \n",
       "2              38B3W  9604/9604/4N2E5.raw_alg.faa.gz,9989/9989/4Q1E1...   \n",
       "3              38B3X  9989/9989/4Q2TC.raw_alg.faa.gz,7214/7214/45XNB...   \n",
       "4              38B3Y  7214/7214/45VNG.raw_alg.faa.gz,9604/9604/4N26V...   \n",
       "\n",
       "                                          OG summary  \\\n",
       "0                    {'Rodentia': 1, 'Hominidae': 1}   \n",
       "1   {'Rodentia': 4, 'Rhabditida': 1, 'Hominidae': 4}   \n",
       "2                    {'Rodentia': 1, 'Hominidae': 1}   \n",
       "3  {'Rodentia': 3, 'Rhabditida': 1, 'Drosophilida...   \n",
       "4  {'Saccharomycetaceae': 1, 'Rhabditida': 1, 'Dr...   \n",
       "\n",
       "                                               Genes  Number of proteins  \\\n",
       "0                                      usf3,kiaa2018                 179   \n",
       "1           ctnnd2,arvcf,jac-1,ctnnd1,cbr-jac-1,pkp4                 546   \n",
       "2                                             fkbp14                 142   \n",
       "3  lamc1,lam-2,lanb2,cbr-lam-2,dyak_lanb2,dere_la...                 464   \n",
       "4                      nsa2,dyak_ip259,w09c5.1,ip259                 387   \n",
       "\n",
       "   Number of taxons  Number of taxonomic ranks  Number of MSA files COG  \n",
       "0               157                          2                    2   K  \n",
       "1               152                          3                    9  TW  \n",
       "2               125                          2                    2   O  \n",
       "3               163                          4                    8   W  \n",
       "4               338                          4                    4   J  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_vectors = pd.read_csv('results/vectors/eukaryotes_vectors.tsv.gz', sep='\\t')\n",
    "\n",
    "display(Markdown('### Sample of constructed orthologous MSA files table for <br /> Saccharomycetaceae, Rhabditida, Drosophilidae, Rodentia, and Hominidae'))\n",
    "df_vectors.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3.2. Main code\n",
    "\n",
    "\n",
    "* We search for lysine deserts in MSA files exactly as before\n",
    "\n",
    "* We do not distinguish whether MSA has lysine desert type *Aligned* or *Anywhere*; we consider MSA with lysine desert type *Aligned* or *Anywhere* as lysine desert containing MSA.\n",
    "\n",
    "* We consider MSA as with lysine cluster of given type when:\n",
    "\n",
    "    * it has lysine desert type *Aligned* or *Anywhere*\n",
    "    * at least 80% of its sequences have lysine cluster of given type (e.g. for lysine cluster anywhere in the sequence value of `Lysine cluster anywhere fraction` has to be at least 80)\n",
    "    \n",
    "* We save the results as tables (see below)\n",
    "\n",
    "---\n",
    "\n",
    "## Vectors with calculated lysine deserts \n",
    "\n",
    "We will search for lysine desert in each row of the eukaryotic vector table (we will check each MSA file). The results with calculated lysine deserts for MSA files within each vector will be saved as tables.\n",
    "\n",
    "\n",
    "### Naming convention\n",
    "\n",
    "Two lysine deserts vector tables will be saved, according to defined minimum length of lysine desert (150 aa or 50% of median MSA sequence). \n",
    "\n",
    "The tables will be saved to the following directories: \n",
    "\n",
    "* `results/vectors/desert_150/eukaryota_desert_vectors.tsv.gz` \n",
    "\n",
    "* `results/vectors/desert_0.5/eukaryota_desert_vectors.tsv.gz` \n",
    "\n",
    "where 150 or 0.5 refers to analyzed minimum length of lysine desert.\n",
    "\n",
    "### Indexing\n",
    "\n",
    "Summary tables are indexed by modified Ophistokonta Orthologous Group ID.\n",
    "\n",
    "### Columns description\n",
    "\n",
    "Summary tables contain the following information:\n",
    "\n",
    "| Column  | Description  |\n",
    "|:--|:--|\n",
    "| **Total taxonomic groups number**  | Total number of selected eukaryotic groups present in the vector  |\n",
    "| **Total MSA number**  |  Total number of MSA files in the vector |\n",
    "|  **Ophistokonta COG** | COG annotation for Ophistokonta Orthologous Group (entire vector); note that individual MSA COG annotations may be diffrent from it  |\n",
    "| ***TAXONOMIC GROUP* total MSA number**  | Total number of MSA files from *TAXONOMIC GROUP* in the vector  |\n",
    "|***TAXONOMIC GROUP* lysine desert MSAs percentage** | Fraction of lysine desert containing MSA files among all MSA files from given *TAXONOMIC GROUP* in the vector|\n",
    "|***TAXONOMIC GROUP* non-lysine desert MSAs percentage** | Fraction of non-lysine desert MSA files among all MSA files from given *TAXONOMIC GROUP* in the vector| \n",
    "|***TAXONOMIC GROUP* lysine desert gene symbols** | Gene Symbols associated with lysine desert containing MSA files in the vector|\n",
    "|***TAXONOMIC GROUP* non-lysine desert gene symbols** | Gene Symbols associated with non-lysine desert MSA files in the vector|\n",
    "|***TAXONOMIC GROUP*  lysine desert MSA COG** | COG annotations of lysine desert MSA files from *TAXONOMIC GROUP* in the vector| \n",
    "|***TAXONOMIC GROUP*  non-lysine desert MSA COG** | COG annotations of non-lysine desert MSA files from *TAXONOMIC GROUP* in the vector| \n",
    "|***TAXONOMIC GROUP* lysine desert MSAs with lysine cluster anywhere percentage** | Percentage of lysine desert with lysine cluster anywhere in sequence MSA files among all lysine desert MSA files from given *TAXONOMIC GROUP* in the vector|\n",
    "|***TAXONOMIC GROUP* lysine desert MSAs with lysine cluster N-end percentage** | Percentage of lysine desert with lysine cluster at N-end MSA files among all lysine desert MSA files from given *TAXONOMIC GROUP* in the vector|\n",
    "|***TAXONOMIC GROUP* lysine desert MSAs with lysine cluster C-end percentage** | Percentage of lysine desert with lysine cluster at C-end MSA files among all lysine desert MSA files from given *TAXONOMIC GROUP* in the vector|\n",
    "|***TAXONOMIC GROUP* longest lysine-less regions medians among lysine desert MSAs** | Medians of longest lysine less regions in each lysine desert MSA file from given *TAXONOMIC GROUP* in the vector|\n",
    "|***TAXONOMIC GROUP* longest lysine-less regions medians among non-lysine desert MSAs** | Medians of longest lysine less regions in each non-lysine desert MSA file from given *TAXONOMIC GROUP* in the vector|\n",
    "|***TAXONOMIC GROUP* means of lysines per sequence in lysine desert MSAs** | Means of number of lysines per sequence in lysine desert MSA file from given *TAXONOMIC GROUP* in the vector |\n",
    "|***TAXONOMIC GROUP* means of lysines per sequence in non-lysine desert MSAs** |Means of number of lysines per sequence in non-lysine desert MSA file from given *TAXONOMIC GROUP* in the vector |\n",
    "|***TAXONOMIC GROUP* means of lysines fraction per sequence in lysine desert MSAs** | Means of fraction of lysines per sequence in lysine desert MSA file from given *TAXONOMIC GROUP* in the vector|\n",
    "|***TAXONOMIC GROUP* means of lysines fraction per sequence in non-lysine desert MSAs** |Means of fraction of lysines per sequence in non-lysine desert MSA file from given *TAXONOMIC GROUP* in the vector |\n",
    "|***TAXONOMIC GROUP* means of arginines per sequence in lysine desert MSAs** |Means of number of aginines per sequence in lysine desert MSA file from given *TAXONOMIC GROUP* in the vector |\n",
    "|***TAXONOMIC GROUP* means of arginines per sequence in non-lysine desert MSAs** | Means of number of aginines per sequence in non-lysine desert MSA file from given *TAXONOMIC GROUP* in the vector|\n",
    "|***TAXONOMIC GROUP* means of arginines fraction per sequence in lysine desert MSAs** | Means of fraction of aginines per sequence in lysine desert MSA file from given *TAXONOMIC GROUP* in the vector|\n",
    "|***TAXONOMIC GROUP* means of arginines fraction per sequence in non-lysine desert MSAs** |  Means of fraction of aginines per sequence in non-lysine desert MSA file from given *TAXONOMIC GROUP* in the vector|\n",
    "|***TAXONOMIC GROUP* paths to lysine desert MSAs** | Paths to lysine desert MSA files from given *TAXONOMIC GROUP* in the vector|\n",
    "|***TAXONOMIC GROUP* paths to non-lysine desert MSAs** | Paths to non-lysine desert MSA files from given *TAXONOMIC GROUP* in the vector|\n",
    "|***TAXONOMIC GROUP* all MSAs sequence length medians** | Medians of MSA sequence length for all MSA files from given *TAXONOMIC GROUP* in the vector|\n",
    "\n",
    "*TAXONOMIC GROUP* = *Saccharomycetaceae/Rhabditida/Drosophilidae/Rodentia/Hominidae*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def search_lysine_desert_vectors(desert_min_length, taxonomic_group):\n",
    "    \"\"\"Find lysine-less regions by building MSA desert conservation profiles for each aln of OG\n",
    "    \n",
    "        :desert_min_lengthh: minimal lysine-less region length or sequence fraction to consider lysine desert\n",
    "        :type minimal_seq_length: int or float\n",
    "        :param taxonomic_group: list of NCBI IDs (compatible to created vector file) to search for lysine desert\n",
    "        :type taxonomic_group: list\n",
    "        :return: dictionary with lists of lysine desert scores for each taxonomic group within each vector\n",
    "        :rtype: dict\n",
    "    \"\"\"\n",
    "    \n",
    "    def vector_division(n, d):\n",
    "        return round((n / d)*100, 2) if d else None\n",
    "    \n",
    "    vectors_lysine_summary = {}\n",
    "\n",
    "    for ident, value in df_vectors.iterrows():\n",
    "        \n",
    "        vectors_lysine_summary[value[0]] = []  \n",
    "        MSA_list = value[1]\n",
    "        \n",
    "        total_MSAs = len(MSA_list.split(','))\n",
    "        Ophistokonta_COG = value[8]\n",
    "        \n",
    "        total_eukaryotic_groups = set()\n",
    "        total_deserts = 0\n",
    "        deserts_summary = {}\n",
    "        \n",
    "        for group in taxonomic_group:\n",
    "            deserts_summary[group] = [0,0,0,[],[],[],[],0,0,0,[],[],[],[],[],[],[],[],[],[],[],[],[]] \n",
    "            '''no of total MSAs,  \n",
    "            no of desert MSAs, no of no desert MSAs, \n",
    "            MSA desert paths, MSA non-desert paths,\n",
    "            gene names desert, gene names non desert,\n",
    "            number of desert MSAs with lysine cluster anywhere, \n",
    "            number of desert MSAs with lysine cluster at N-end, \n",
    "            number of desert MSAs with lysine cluster at C-end,\n",
    "            medians of lysine-less region sequence percentages for desert MSAs,\n",
    "            medians of lysine-less region sequence percentages for non desert MSAs,\n",
    "            means of number of lysines per sequence in desert MSAs, means of number of lysines per sequence in non-desert MSAs,\n",
    "            means of fraction of lysines per sequence in desert MSAs, means of fraction of lysines per sequence in non-desert MSAs,\n",
    "            means of number of arginines per sequence in desert MSAs, means of number of arginines per sequence in non-desert MSAs,\n",
    "            means of fraction of arginines per sequence in desert MSAs, means of fraction of arginines per sequence in non-desert MSAs,\n",
    "            COG annotations of desert MSAs, COG annotations of non-desert MSAs, \n",
    "            medians of protein length\n",
    "            '''\n",
    "        \n",
    "        for msa in MSA_list.split(','):\n",
    "            \n",
    "            taxid = msa.split('/')[0]\n",
    "            MSA_id = msa.split('/')[-1].split('.')[0]\n",
    "            total_eukaryotic_groups.add(taxid)\n",
    "            \n",
    "            deserts_summary[taxid][0] += 1 # add MSA to total number of MSA files from this taxonomic group\n",
    "\n",
    "            aln = FindDesertMSA('data/{}/{}/{}.raw_alg.faa.gz'.format(taxid, taxid, MSA_id))\n",
    "            \n",
    "            \n",
    "            # Add median of sequence length\n",
    "            deserts_summary[taxid][22].append(str(aln.calculate_median_seq_length())) \n",
    "            \n",
    "            # Check if MSA has lysine desert\n",
    "            scores = aln.find_deserts(desert_min_length)\n",
    "\n",
    "            # MSA has lysine desert\n",
    "            if len(scores) > 0:\n",
    "                deserts_summary[taxid][1] += 1\n",
    "                deserts_summary[taxid][3].append('data/{}/{}/{}.raw_alg.faa.gz'.format(taxid, taxid, MSA_id))\n",
    "                deserts_summary[taxid][5].extend(aln.get_gene_names())\n",
    "                deserts_summary[taxid][10].append(str(aln.calc_median_lysine_less_region_percentage()))\n",
    "\n",
    "                scores = scores[0]\n",
    "                # Lysine cluster percentages - how many MSAs in vector of given taxonomic groups have lysine cluster \n",
    "                # We classify MSA as with lysine cluster if 80% sequences (default value) have lysine cluster \n",
    "                lysine_cluster_anywhere, lysine_cluster_N_end, lysine_cluster_C_end = scores[14], scores[15], scores[16]\n",
    "                if lysine_cluster_anywhere >= LYSINE_CLUSTER_MSA_THRESHOLD_VECTORS:\n",
    "                    deserts_summary[taxid][7] += 1\n",
    "                if lysine_cluster_N_end >= LYSINE_CLUSTER_MSA_THRESHOLD_VECTORS:\n",
    "                    deserts_summary[taxid][8] += 1\n",
    "                if lysine_cluster_C_end >= LYSINE_CLUSTER_MSA_THRESHOLD_VECTORS:\n",
    "                    deserts_summary[taxid][9] += 1\n",
    "                    \n",
    "                # Add avg lysine and arginine number per sequence\n",
    "                lys_no, lys_frac = aln.calculate_avg_aa_per_sequence('K')\n",
    "                deserts_summary[taxid][12].append(str(lys_no))\n",
    "                deserts_summary[taxid][14].append(str(lys_frac))\n",
    "                \n",
    "                arg_no, arg_frac = aln.calculate_avg_aa_per_sequence('R')\n",
    "                deserts_summary[taxid][16].append(str(arg_no))\n",
    "                deserts_summary[taxid][18].append(str(arg_frac))\n",
    "                \n",
    "                # Add COG annotation\n",
    "                deserts_summary[taxid][20].extend(aln.get_COG())\n",
    "            \n",
    "            \n",
    "            else: # MSA does not have lysine desert\n",
    "                deserts_summary[taxid][2] += 1\n",
    "                deserts_summary[taxid][4].append('data/{}/{}/{}.raw_alg.faa.gz'.format(taxid, taxid, MSA_id))\n",
    "                deserts_summary[taxid][6].extend(aln.get_gene_names())\n",
    "                deserts_summary[taxid][11].append(str(aln.calc_median_lysine_less_region_percentage()))\n",
    "                \n",
    "                # Add avg lysine and arginine number per sequence\n",
    "                lys_no, lys_frac = aln.calculate_avg_aa_per_sequence('K')\n",
    "                deserts_summary[taxid][13].append(str(lys_no))\n",
    "                deserts_summary[taxid][15].append(str(lys_frac))\n",
    "                \n",
    "                arg_no, arg_frac = aln.calculate_avg_aa_per_sequence('R')\n",
    "                deserts_summary[taxid][17].append(str(arg_no))\n",
    "                deserts_summary[taxid][19].append(str(arg_frac))\n",
    "                \n",
    "                # Add COG annotation\n",
    "                deserts_summary[taxid][21].extend(aln.get_COG())\n",
    "        \n",
    "        # Save results\n",
    "        vectors_lysine_summary[value[0]].extend([len(total_eukaryotic_groups), total_MSAs, Ophistokonta_COG])\n",
    "        \n",
    "        for group in taxonomic_group:\n",
    "            # If taxonomic group did not occur in vector\n",
    "            if deserts_summary[group][0] == 0: \n",
    "                vectors_lysine_summary[value[0]].extend([None]*23) # we save 23 columns with different info for each taxonomic group\n",
    "            \n",
    "            else:\n",
    "\n",
    "                lysine_cluster_anywhere_percentage = vector_division(deserts_summary[group][7], deserts_summary[group][1])\n",
    "                lysine_cluster_N_end_percentage = vector_division(deserts_summary[group][8], deserts_summary[group][1])\n",
    "                lysine_cluster_C_end_percentage = vector_division(deserts_summary[group][9], deserts_summary[group][1])\n",
    "                \n",
    "                vectors_lysine_summary[value[0]].extend([deserts_summary[group][0],\n",
    "                                                 round((deserts_summary[group][1]/deserts_summary[group][0])*100, 2),\n",
    "                                                 round((deserts_summary[group][2]/deserts_summary[group][0])*100, 2),\n",
    "                                                 ';'.join(deserts_summary[group][5]), ';'.join(deserts_summary[group][6]), \n",
    "                                                 ';'.join(deserts_summary[group][20]), ';'.join(deserts_summary[group][21]),\n",
    "                                                 lysine_cluster_anywhere_percentage,\n",
    "                                                 lysine_cluster_N_end_percentage,\n",
    "                                                 lysine_cluster_C_end_percentage,\n",
    "                                                 ';'.join(deserts_summary[group][10]), ';'.join(deserts_summary[group][11]),\n",
    "                                                 ';'.join(deserts_summary[group][12]), ';'.join(deserts_summary[group][13]),\n",
    "                                                 ';'.join(deserts_summary[group][14]), ';'.join(deserts_summary[group][15]),\n",
    "                                                 ';'.join(deserts_summary[group][16]), ';'.join(deserts_summary[group][17]),\n",
    "                                                 ';'.join(deserts_summary[group][18]), ';'.join(deserts_summary[group][19]),\n",
    "                                                 ';'.join(deserts_summary[group][3]), ';'.join(deserts_summary[group][4]),\n",
    "                                                 ';'.join(deserts_summary[group][22])])\n",
    "                                            \n",
    "    return vectors_lysine_summary"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3.3. Run analysis for eukaryotes "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_analysis(taxonomic_group, out_name):\n",
    "    \n",
    "    for desert_length in desert_min_lengths:\n",
    "        \n",
    "        result = search_lysine_desert_vectors(desert_length, eukaryota)\n",
    "        df = pd.DataFrame.from_dict(result, orient='index')\n",
    "        df.index.name = 'Ophistokonta OG ID'\n",
    "        #display(df)\n",
    "        \n",
    "        columns = ['Total taxonomic groups number', 'Total MSA number', 'Ophistokonta COG']\n",
    "        common_col_names = ['total MSA number', \n",
    "                            'lysine desert MSAs percentage', 'non-lysine desert MSAs percentage', \n",
    "                            'lysine desert gene symbols', 'non-lysine desert gene symbols', \n",
    "                            'lysine desert MSA COG', 'non-lysine desert MSA COG',\n",
    "                            'lysine desert MSAs with lysine cluster anywhere percentage',\n",
    "                            'lysine desert MSAs with lysine cluster N-end percentage',\n",
    "                            'lysine desert MSAs with lysine cluster C-end percentage',\n",
    "                            'longest lysine-less regions medians among lysine desert MSAs',\n",
    "                            'longest lysine-less regions medians among non-lysine desert MSAs',\n",
    "                            'means of lysines per sequence in lysine desert MSAs',\n",
    "                            'means of lysines per sequence in non-lysine desert MSAs',\n",
    "                            'means of lysines fraction per sequence in lysine desert MSAs',\n",
    "                            'means of lysines fraction per sequence in non-lysine desert MSAs',\n",
    "                            'means of arginines per sequence in lysine desert MSAs',\n",
    "                            'means of arginines per sequence in non-lysine desert MSAs',\n",
    "                            'means of arginines fraction per sequence in lysine desert MSAs',\n",
    "                            'means of arginines fraction per sequence in non-lysine desert MSAs',\n",
    "                            'paths to lysine desert MSAs', 'paths to non-lysine desert MSAs',\n",
    "                            'all MSAs sequence length medians']\n",
    "        \n",
    "        for group in taxonomic_group:\n",
    "            group_name = ncbi.get_taxid_translator([group])[int(group)]\n",
    "            for col_name in common_col_names:\n",
    "                columns.append('{} {}'.format(group_name, col_name))\n",
    "        \n",
    "        df.columns = columns\n",
    "        \n",
    "        vectors_save_path = 'results/vectors/desert_{}'.format(desert_length)\n",
    "        if not os.path.exists(vectors_save_path):\n",
    "            os.makedirs(vectors_save_path)\n",
    "            \n",
    "        df.to_csv('results/vectors/desert_{}/{}_desert_vectors.tsv.gz'.format(desert_length, out_name), \n",
    "                   sep='\\t', compression='gzip')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "run_analysis(eukaryota, 'eukaryota')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4. Analyze results from vectors \n",
    "\n",
    "Analyze results for vectors for **lysine desert of minimum 50% median MSA sequence length**."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Find highly conserved lysine deserts\n",
    "\n",
    "Select lysine desert min. 50% conserved in all MSA files from *Drosophilidae, Rodentia*, and *Hominidae* (in *Saccharomycetaceae* and *Rhabditida*, MSAs were either absent or, if present, conserved lysine desert min. 50% also needed to occur)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/63/q2_r09gx5q1_k4n6r4pwry5w0000gn/T/ipykernel_19862/664864460.py:3: DtypeWarning: Columns (60,62,64,66,68,83,85,87,89,91) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv('results/vectors/desert_0.5/eukaryota_desert_vectors.tsv.gz', sep='\\t')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Saccharomycetaceae lysine desert gene symbols</th>\n",
       "      <th>Saccharomycetaceae paths to lysine desert MSAs</th>\n",
       "      <th>Rhabditida lysine desert gene symbols</th>\n",
       "      <th>Rhabditida paths to lysine desert MSAs</th>\n",
       "      <th>Drosophilidae lysine desert gene symbols</th>\n",
       "      <th>Drosophilidae paths to lysine desert MSAs</th>\n",
       "      <th>Rodentia lysine desert gene symbols</th>\n",
       "      <th>Rodentia paths to lysine desert MSAs</th>\n",
       "      <th>Hominidae lysine desert gene symbols</th>\n",
       "      <th>Hominidae paths to lysine desert MSAs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1100</th>\n",
       "      <td>rad23</td>\n",
       "      <td>data/4893/4893/3RYJU.raw_alg.faa.gz</td>\n",
       "      <td>rad-23</td>\n",
       "      <td>data/6236/6236/40YDM.raw_alg.faa.gz</td>\n",
       "      <td>cg10694;rad23</td>\n",
       "      <td>data/7214/7214/45QJE.raw_alg.faa.gz;data/7214/7214/45S3J.raw_alg.faa.gz</td>\n",
       "      <td>rad23a;rad23b</td>\n",
       "      <td>data/9989/9989/4Q4U8.raw_alg.faa.gz;data/9989/9989/4PVU6.raw_alg.faa.gz</td>\n",
       "      <td>rad23b;rad23a</td>\n",
       "      <td>data/9604/9604/4N3HT.raw_alg.faa.gz;data/9604/9604/4MW4P.raw_alg.faa.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1149</th>\n",
       "      <td>dsk2</td>\n",
       "      <td>data/4893/4893/3RZAV.raw_alg.faa.gz</td>\n",
       "      <td>ubql-1</td>\n",
       "      <td>data/6236/6236/40XXY.raw_alg.faa.gz</td>\n",
       "      <td>ubqn</td>\n",
       "      <td>data/7214/7214/45N1A.raw_alg.faa.gz</td>\n",
       "      <td>ubqln1;ubqln2</td>\n",
       "      <td>data/9989/9989/4PU7Q.raw_alg.faa.gz;data/9989/9989/4PX5U.raw_alg.faa.gz</td>\n",
       "      <td>ubqln4;ubqln1</td>\n",
       "      <td>data/9604/9604/4MZQ3.raw_alg.faa.gz;data/9604/9604/4N4BU.raw_alg.faa.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2135</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ssdp</td>\n",
       "      <td>data/7214/7214/45MWY.raw_alg.faa.gz</td>\n",
       "      <td>ssbp2;ssbp3</td>\n",
       "      <td>data/9989/9989/4Q3QK.raw_alg.faa.gz;data/9989/9989/4Q83H.raw_alg.faa.gz</td>\n",
       "      <td>ssbp3;ssbp2</td>\n",
       "      <td>data/9604/9604/4N31C.raw_alg.faa.gz;data/9604/9604/4N2YF.raw_alg.faa.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3262</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>data/7214/7214/45WN3.raw_alg.faa.gz</td>\n",
       "      <td>bag6</td>\n",
       "      <td>data/9989/9989/4Q2H9.raw_alg.faa.gz</td>\n",
       "      <td>bag6</td>\n",
       "      <td>data/9604/9604/4MSAK.raw_alg.faa.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3274</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cg11982</td>\n",
       "      <td>data/7214/7214/45MZB.raw_alg.faa.gz</td>\n",
       "      <td>rnf126;rnf115</td>\n",
       "      <td>data/9989/9989/4PW0V.raw_alg.faa.gz;data/9989/9989/4Q1XU.raw_alg.faa.gz</td>\n",
       "      <td>rnf115;rnf126</td>\n",
       "      <td>data/9604/9604/4MS52.raw_alg.faa.gz;data/9604/9604/4N3NE.raw_alg.faa.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6582</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tag-353</td>\n",
       "      <td>data/6236/6236/40S1S.raw_alg.faa.gz</td>\n",
       "      <td>herp</td>\n",
       "      <td>data/7214/7214/45TNG.raw_alg.faa.gz</td>\n",
       "      <td>herpud2;herpud1</td>\n",
       "      <td>data/9989/9989/4PS71.raw_alg.faa.gz;data/9989/9989/4PZKA.raw_alg.faa.gz</td>\n",
       "      <td>herpud2;herpud1</td>\n",
       "      <td>data/9604/9604/4N000.raw_alg.faa.gz;data/9604/9604/4MYAQ.raw_alg.faa.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7004</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>data/7214/7214/45QY7.raw_alg.faa.gz</td>\n",
       "      <td>ubl7</td>\n",
       "      <td>data/9989/9989/4PSW9.raw_alg.faa.gz</td>\n",
       "      <td>ubl7</td>\n",
       "      <td>data/9604/9604/4MUZ5.raw_alg.faa.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7706</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>toe-4</td>\n",
       "      <td>data/6236/6236/4108P.raw_alg.faa.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>data/7214/7214/45SZV.raw_alg.faa.gz</td>\n",
       "      <td>rnf111;rnf165</td>\n",
       "      <td>data/9989/9989/4PZMH.raw_alg.faa.gz;data/9989/9989/4PZMF.raw_alg.faa.gz</td>\n",
       "      <td>rnf165;rnf111</td>\n",
       "      <td>data/9604/9604/4N6CW.raw_alg.faa.gz;data/9604/9604/4MZ0F.raw_alg.faa.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10225</th>\n",
       "      <td>mix17</td>\n",
       "      <td>data/4893/4893/3S17Q.raw_alg.faa.gz</td>\n",
       "      <td>har-1</td>\n",
       "      <td>data/6236/6236/411SE.raw_alg.faa.gz</td>\n",
       "      <td>NaN</td>\n",
       "      <td>data/7214/7214/45TPG.raw_alg.faa.gz</td>\n",
       "      <td>chchd2</td>\n",
       "      <td>data/9989/9989/4Q5AB.raw_alg.faa.gz</td>\n",
       "      <td>chchd2</td>\n",
       "      <td>data/9604/9604/4MVN7.raw_alg.faa.gz</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Saccharomycetaceae lysine desert gene symbols  \\\n",
       "1100                                          rad23   \n",
       "1149                                           dsk2   \n",
       "2135                                            NaN   \n",
       "3262                                            NaN   \n",
       "3274                                            NaN   \n",
       "6582                                            NaN   \n",
       "7004                                            NaN   \n",
       "7706                                            NaN   \n",
       "10225                                         mix17   \n",
       "\n",
       "      Saccharomycetaceae paths to lysine desert MSAs  \\\n",
       "1100             data/4893/4893/3RYJU.raw_alg.faa.gz   \n",
       "1149             data/4893/4893/3RZAV.raw_alg.faa.gz   \n",
       "2135                                             NaN   \n",
       "3262                                             NaN   \n",
       "3274                                             NaN   \n",
       "6582                                             NaN   \n",
       "7004                                             NaN   \n",
       "7706                                             NaN   \n",
       "10225            data/4893/4893/3S17Q.raw_alg.faa.gz   \n",
       "\n",
       "      Rhabditida lysine desert gene symbols  \\\n",
       "1100                                 rad-23   \n",
       "1149                                 ubql-1   \n",
       "2135                                    NaN   \n",
       "3262                                    NaN   \n",
       "3274                                    NaN   \n",
       "6582                                tag-353   \n",
       "7004                                    NaN   \n",
       "7706                                  toe-4   \n",
       "10225                                 har-1   \n",
       "\n",
       "      Rhabditida paths to lysine desert MSAs  \\\n",
       "1100     data/6236/6236/40YDM.raw_alg.faa.gz   \n",
       "1149     data/6236/6236/40XXY.raw_alg.faa.gz   \n",
       "2135                                     NaN   \n",
       "3262                                     NaN   \n",
       "3274                                     NaN   \n",
       "6582     data/6236/6236/40S1S.raw_alg.faa.gz   \n",
       "7004                                     NaN   \n",
       "7706     data/6236/6236/4108P.raw_alg.faa.gz   \n",
       "10225    data/6236/6236/411SE.raw_alg.faa.gz   \n",
       "\n",
       "      Drosophilidae lysine desert gene symbols  \\\n",
       "1100                             cg10694;rad23   \n",
       "1149                                      ubqn   \n",
       "2135                                      ssdp   \n",
       "3262                                       NaN   \n",
       "3274                                   cg11982   \n",
       "6582                                      herp   \n",
       "7004                                       NaN   \n",
       "7706                                       NaN   \n",
       "10225                                      NaN   \n",
       "\n",
       "                                     Drosophilidae paths to lysine desert MSAs  \\\n",
       "1100   data/7214/7214/45QJE.raw_alg.faa.gz;data/7214/7214/45S3J.raw_alg.faa.gz   \n",
       "1149                                       data/7214/7214/45N1A.raw_alg.faa.gz   \n",
       "2135                                       data/7214/7214/45MWY.raw_alg.faa.gz   \n",
       "3262                                       data/7214/7214/45WN3.raw_alg.faa.gz   \n",
       "3274                                       data/7214/7214/45MZB.raw_alg.faa.gz   \n",
       "6582                                       data/7214/7214/45TNG.raw_alg.faa.gz   \n",
       "7004                                       data/7214/7214/45QY7.raw_alg.faa.gz   \n",
       "7706                                       data/7214/7214/45SZV.raw_alg.faa.gz   \n",
       "10225                                      data/7214/7214/45TPG.raw_alg.faa.gz   \n",
       "\n",
       "      Rodentia lysine desert gene symbols  \\\n",
       "1100                        rad23a;rad23b   \n",
       "1149                        ubqln1;ubqln2   \n",
       "2135                          ssbp2;ssbp3   \n",
       "3262                                 bag6   \n",
       "3274                        rnf126;rnf115   \n",
       "6582                      herpud2;herpud1   \n",
       "7004                                 ubl7   \n",
       "7706                        rnf111;rnf165   \n",
       "10225                              chchd2   \n",
       "\n",
       "                                          Rodentia paths to lysine desert MSAs  \\\n",
       "1100   data/9989/9989/4Q4U8.raw_alg.faa.gz;data/9989/9989/4PVU6.raw_alg.faa.gz   \n",
       "1149   data/9989/9989/4PU7Q.raw_alg.faa.gz;data/9989/9989/4PX5U.raw_alg.faa.gz   \n",
       "2135   data/9989/9989/4Q3QK.raw_alg.faa.gz;data/9989/9989/4Q83H.raw_alg.faa.gz   \n",
       "3262                                       data/9989/9989/4Q2H9.raw_alg.faa.gz   \n",
       "3274   data/9989/9989/4PW0V.raw_alg.faa.gz;data/9989/9989/4Q1XU.raw_alg.faa.gz   \n",
       "6582   data/9989/9989/4PS71.raw_alg.faa.gz;data/9989/9989/4PZKA.raw_alg.faa.gz   \n",
       "7004                                       data/9989/9989/4PSW9.raw_alg.faa.gz   \n",
       "7706   data/9989/9989/4PZMH.raw_alg.faa.gz;data/9989/9989/4PZMF.raw_alg.faa.gz   \n",
       "10225                                      data/9989/9989/4Q5AB.raw_alg.faa.gz   \n",
       "\n",
       "      Hominidae lysine desert gene symbols  \\\n",
       "1100                         rad23b;rad23a   \n",
       "1149                         ubqln4;ubqln1   \n",
       "2135                           ssbp3;ssbp2   \n",
       "3262                                  bag6   \n",
       "3274                         rnf115;rnf126   \n",
       "6582                       herpud2;herpud1   \n",
       "7004                                  ubl7   \n",
       "7706                         rnf165;rnf111   \n",
       "10225                               chchd2   \n",
       "\n",
       "                                         Hominidae paths to lysine desert MSAs  \n",
       "1100   data/9604/9604/4N3HT.raw_alg.faa.gz;data/9604/9604/4MW4P.raw_alg.faa.gz  \n",
       "1149   data/9604/9604/4MZQ3.raw_alg.faa.gz;data/9604/9604/4N4BU.raw_alg.faa.gz  \n",
       "2135   data/9604/9604/4N31C.raw_alg.faa.gz;data/9604/9604/4N2YF.raw_alg.faa.gz  \n",
       "3262                                       data/9604/9604/4MSAK.raw_alg.faa.gz  \n",
       "3274   data/9604/9604/4MS52.raw_alg.faa.gz;data/9604/9604/4N3NE.raw_alg.faa.gz  \n",
       "6582   data/9604/9604/4N000.raw_alg.faa.gz;data/9604/9604/4MYAQ.raw_alg.faa.gz  \n",
       "7004                                       data/9604/9604/4MUZ5.raw_alg.faa.gz  \n",
       "7706   data/9604/9604/4N6CW.raw_alg.faa.gz;data/9604/9604/4MZ0F.raw_alg.faa.gz  \n",
       "10225                                      data/9604/9604/4MVN7.raw_alg.faa.gz  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pd.set_option('display.max_colwidth', None)\n",
    "\n",
    "df = pd.read_csv('results/vectors/desert_0.5/eukaryota_desert_vectors.tsv.gz', sep='\\t')\n",
    "\n",
    "# all 3 taxonomic groups must be present in table row and all their MSA files have to contain lysine desert\n",
    "df = df[(df['Drosophilidae lysine desert MSAs percentage'] == 100) & \n",
    "        (df['Rodentia lysine desert MSAs percentage'] == 100) &\n",
    "        (df['Hominidae lysine desert MSAs percentage'] == 100)\n",
    "       ]\n",
    "\n",
    "df['Saccharomycetaceae lysine desert MSAs percentage'] = df['Saccharomycetaceae lysine desert MSAs percentage'].replace(np.nan, 101, regex=True)\n",
    "df['Rhabditida lysine desert MSAs percentage'] = df['Rhabditida lysine desert MSAs percentage'].replace(np.nan, 101, regex=True)\n",
    "df = df[df['Saccharomycetaceae lysine desert MSAs percentage'] >= 100]\n",
    "df = df[df['Rhabditida lysine desert MSAs percentage'] >= 100]\n",
    "df['Saccharomycetaceae lysine desert MSAs percentage'] = df['Saccharomycetaceae lysine desert MSAs percentage'].replace(101, np.nan, regex=True)\n",
    "df['Rhabditida lysine desert MSAs percentage'] = df['Rhabditida lysine desert MSAs percentage'].replace(101, np.nan, regex=True)\n",
    "\n",
    "display(df[[\n",
    "            'Saccharomycetaceae lysine desert gene symbols', 'Saccharomycetaceae paths to lysine desert MSAs',\n",
    "             'Rhabditida lysine desert gene symbols', 'Rhabditida paths to lysine desert MSAs', \n",
    "            'Drosophilidae lysine desert gene symbols', 'Drosophilidae paths to lysine desert MSAs',\n",
    "             'Rodentia lysine desert gene symbols', 'Rodentia paths to lysine desert MSAs',\n",
    "             'Hominidae lysine desert gene symbols', 'Hominidae paths to lysine desert MSAs'\n",
    "            ]])\n",
    "\n",
    "df.to_csv('results/supplementary/Table_lysine_desert_0.5_conserved_eukaryotic_vectors.tsv', sep='\\t')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
