{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prepare data for the detailed analysis of lysine deserts in the human  proteome\n",
    "\n",
    "\n",
    "<br />\n",
    "\n",
    "This Jupyter Notebook, except cells in par. 4 & 6, may be skipped, so the authors' preprocessed data will be used in the subsequent analyses."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Import libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Download data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "mkdir -p data\n",
    "mkdir -p results\n",
    "mkdir -p results/TMH_predictions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.1. Download UniProt data\n",
    "\n",
    "Directly on the UniProt web site, download all human sequences with the following information (**keep the indicated order!**):\n",
    "\n",
    "- Entry\t\n",
    "- Reviewed\t\n",
    "- Entry Name\t\n",
    "- Protein names\t\n",
    "- Gene Names\t\n",
    "- Organism\t\n",
    "- Length\t\n",
    "- Proteomes\t\n",
    "- Subcellular location [CC]\t\n",
    "- Gene Ontology (biological process)\t\n",
    "- Gene Ontology (cellular component)\t\n",
    "- Gene Ontology (molecular function)\t\n",
    "- Sequence\t\n",
    "- Function [CC]\t\n",
    "- Tissue specificity\n",
    "- Pathway\t\n",
    "- Involvement in disease\t\n",
    "- Natural variant\t\n",
    "- Protein families\t\n",
    "- Interacts with\t\n",
    "- Compositional bias\t\n",
    "- 3D\t\n",
    "- Protein existence\t\n",
    "- Domain [FT]\n",
    "\n",
    "and save in `data` as `Uniprot_raw_human_proteome.tsv`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.2. Add data on disordered regions from the MobiDB\n",
    "\n",
    "Download dataset of all human intrinsically disordered regions (IDRs) from [MobiDB](https://mobidb.bio.unipd.it) and save as `files/MobiDB_human_IDRs.tsv`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "response = requests.get(\"https://mobidb.bio.unipd.it/api/download?ncbi_taxon_id=9606&format=tsv\")\n",
    "\n",
    "with open('data/MobiDB_human_IDRs.tsv','w') as fd:\n",
    "    fd.write(response.text)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "mdb = pd.read_csv('data/MobiDB_human_IDRs.tsv', sep='\\t')\n",
    "\n",
    "# from MobiDB datase, select pLDDT-based predictions from the AlphaFold2\n",
    "mdb = mdb[mdb['feature'] == 'prediction-plddt-alphafold']\n",
    "mdb.drop(['feature', 'content_count'], axis=1, inplace=True)\n",
    "\n",
    "mdb.rename(columns={'acc':'Entry', \n",
    "                    'start..end': 'Struct_regions_start..end',\n",
    "                    'content_fraction': 'Struct_regions_fraction'}, inplace=True)\n",
    "\n",
    "# Get ranges of IDRs\n",
    "IDRs = []\n",
    "IDRs_fraction = []\n",
    "\n",
    "for key, value in mdb.iterrows():\n",
    "    length = value[3]\n",
    "    IDRs_fraction.append(round((1.0 - float(value[2]))*100, 3))\n",
    "    struct_ranges = value[1].split(',')\n",
    "    far_left = 0\n",
    "    prot_IDRs = []\n",
    "    \n",
    "    for i in range(len(struct_ranges)):\n",
    "        start = int(struct_ranges[i].split('..')[0])\n",
    "        end = int(struct_ranges[i].split('..')[1])\n",
    "        min_r = 999999999\n",
    "        max_r = 0\n",
    "        for el in range(far_left, end+1):\n",
    "            if el not in range(start, end+1):\n",
    "                if el < min_r:\n",
    "                    min_r = el\n",
    "                if el > max_r:\n",
    "                    max_r = el\n",
    "        \n",
    "        to_append = f'{min_r}..{max_r}'\n",
    "        if to_append != '0..0':\n",
    "            prot_IDRs.append(to_append)\n",
    "        far_left = end + 1\n",
    "        \n",
    "    end_to_append = f'{end+1}..{length}'\n",
    "    if end_to_append != f'{length}..{length}' and end+1 <= length:\n",
    "        prot_IDRs.append(end_to_append)\n",
    "    else:\n",
    "        prot_IDRs.append(f'{end+1}')\n",
    "        \n",
    "    IDRs.append(','.join(prot_IDRs))\n",
    "\n",
    "mdb['IDR_region_start..end'] = IDRs\n",
    "mdb['IDR_region_fraction'] = IDRs_fraction\n",
    "mdb['Struct_regions_fraction'] = mdb['Struct_regions_fraction'] *100\n",
    "mdb.drop('length', axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.3. Merge data from the UniProt with the MobiDB information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ns/opt/anaconda3/envs/Lys-deserts/lib/python3.9/site-packages/pandas/core/internals/blocks.py:2199: RuntimeWarning: invalid value encountered in cast\n",
      "  values = values.astype(str)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Reviewed</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Protein names</th>\n",
       "      <th>Gene Names</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Length</th>\n",
       "      <th>Proteomes</th>\n",
       "      <th>Subcellular location [CC]</th>\n",
       "      <th>Gene Ontology (biological process)</th>\n",
       "      <th>...</th>\n",
       "      <th>Protein families</th>\n",
       "      <th>Interacts with</th>\n",
       "      <th>Compositional bias</th>\n",
       "      <th>3D</th>\n",
       "      <th>Protein existence</th>\n",
       "      <th>Domain [FT]</th>\n",
       "      <th>Struct_regions_start..end</th>\n",
       "      <th>Struct_regions_fraction</th>\n",
       "      <th>IDR_region_start..end</th>\n",
       "      <th>IDR_region_fraction</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A087X1C5</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>CP2D7_HUMAN</td>\n",
       "      <td>Putative cytochrome P450 2D7 (EC 1.14.14.1)</td>\n",
       "      <td>CYP2D7</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>515.0</td>\n",
       "      <td>UP000005640: Chromosome 22</td>\n",
       "      <td>SUBCELLULAR LOCATION: Membrane {ECO:0000305}; ...</td>\n",
       "      <td>arachidonic acid metabolic process [GO:0019369...</td>\n",
       "      <td>...</td>\n",
       "      <td>Cytochrome P450 family</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Uncertain</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4..142,148..227,229..329,348..515</td>\n",
       "      <td>94.8</td>\n",
       "      <td>0..3,143..147,228..228,330..347,516</td>\n",
       "      <td>5.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A0B4J2F0</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>PIOS1_HUMAN</td>\n",
       "      <td>Protein PIGBOS1 (PIGB opposite strand protein 1)</td>\n",
       "      <td>PIGBOS1</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>54.0</td>\n",
       "      <td>UP000005640: Chromosome 15</td>\n",
       "      <td>SUBCELLULAR LOCATION: Mitochondrion outer memb...</td>\n",
       "      <td>regulation of endoplasmic reticulum unfolded p...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Q96S66</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Evidence at protein level</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6..53</td>\n",
       "      <td>90.7</td>\n",
       "      <td>0..5,54</td>\n",
       "      <td>9.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A0B4J2F2</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>SIK1B_HUMAN</td>\n",
       "      <td>Putative serine/threonine-protein kinase SIK1B...</td>\n",
       "      <td>SIK1B</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>783.0</td>\n",
       "      <td>UP000005640: Chromosome 21</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cellular response to glucose starvation [GO:00...</td>\n",
       "      <td>...</td>\n",
       "      <td>Protein kinase superfamily, CAMK Ser/Thr prote...</td>\n",
       "      <td>Q9BRT9; Q68G74; Q14696; P78337; Q9NRG1; P62072</td>\n",
       "      <td>COMPBIAS 457..477; /note=\"Polar residues\"; /ev...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Uncertain</td>\n",
       "      <td>DOMAIN 27..278; /note=\"Protein kinase\"; /evide...</td>\n",
       "      <td>21..280,303..352</td>\n",
       "      <td>39.6</td>\n",
       "      <td>0..20,281..302,353..783</td>\n",
       "      <td>60.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A0C5B5G6</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>MOTSC_HUMAN</td>\n",
       "      <td>Mitochondrial-derived peptide MOTS-c (Mitochon...</td>\n",
       "      <td>MT-RNR1</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>16.0</td>\n",
       "      <td>UP000005640: Mitochondrion</td>\n",
       "      <td>SUBCELLULAR LOCATION: Secreted {ECO:0000269|Pu...</td>\n",
       "      <td>activation of protein kinase activity [GO:0032...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Evidence at protein level</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6..15</td>\n",
       "      <td>62.5</td>\n",
       "      <td>0..5,16</td>\n",
       "      <td>37.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A0K2S4Q6</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>CD3CH_HUMAN</td>\n",
       "      <td>Protein CD300H (CD300 antigen-like family memb...</td>\n",
       "      <td>CD300H</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>201.0</td>\n",
       "      <td>UP000005640: Chromosome 17</td>\n",
       "      <td>SUBCELLULAR LOCATION: [Isoform 1]: Membrane {E...</td>\n",
       "      <td>neutrophil chemotaxis [GO:0030593]</td>\n",
       "      <td>...</td>\n",
       "      <td>CD300 family</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Evidence at protein level</td>\n",
       "      <td>DOMAIN 25..123; /note=\"Ig-like V-type\"; /evide...</td>\n",
       "      <td>24..62,81..115,127..139,174..193</td>\n",
       "      <td>55.2</td>\n",
       "      <td>0..23,63..80,116..126,140..173,194..201</td>\n",
       "      <td>44.8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Entry  Reviewed   Entry Name  \\\n",
       "0  A0A087X1C5  reviewed  CP2D7_HUMAN   \n",
       "1  A0A0B4J2F0  reviewed  PIOS1_HUMAN   \n",
       "2  A0A0B4J2F2  reviewed  SIK1B_HUMAN   \n",
       "3  A0A0C5B5G6  reviewed  MOTSC_HUMAN   \n",
       "4  A0A0K2S4Q6  reviewed  CD3CH_HUMAN   \n",
       "\n",
       "                                       Protein names Gene Names  \\\n",
       "0        Putative cytochrome P450 2D7 (EC 1.14.14.1)     CYP2D7   \n",
       "1   Protein PIGBOS1 (PIGB opposite strand protein 1)    PIGBOS1   \n",
       "2  Putative serine/threonine-protein kinase SIK1B...      SIK1B   \n",
       "3  Mitochondrial-derived peptide MOTS-c (Mitochon...    MT-RNR1   \n",
       "4  Protein CD300H (CD300 antigen-like family memb...     CD300H   \n",
       "\n",
       "               Organism  Length                   Proteomes  \\\n",
       "0  Homo sapiens (Human)   515.0  UP000005640: Chromosome 22   \n",
       "1  Homo sapiens (Human)    54.0  UP000005640: Chromosome 15   \n",
       "2  Homo sapiens (Human)   783.0  UP000005640: Chromosome 21   \n",
       "3  Homo sapiens (Human)    16.0  UP000005640: Mitochondrion   \n",
       "4  Homo sapiens (Human)   201.0  UP000005640: Chromosome 17   \n",
       "\n",
       "                           Subcellular location [CC]  \\\n",
       "0  SUBCELLULAR LOCATION: Membrane {ECO:0000305}; ...   \n",
       "1  SUBCELLULAR LOCATION: Mitochondrion outer memb...   \n",
       "2                                                NaN   \n",
       "3  SUBCELLULAR LOCATION: Secreted {ECO:0000269|Pu...   \n",
       "4  SUBCELLULAR LOCATION: [Isoform 1]: Membrane {E...   \n",
       "\n",
       "                  Gene Ontology (biological process)  ...  \\\n",
       "0  arachidonic acid metabolic process [GO:0019369...  ...   \n",
       "1  regulation of endoplasmic reticulum unfolded p...  ...   \n",
       "2  cellular response to glucose starvation [GO:00...  ...   \n",
       "3  activation of protein kinase activity [GO:0032...  ...   \n",
       "4                 neutrophil chemotaxis [GO:0030593]  ...   \n",
       "\n",
       "                                    Protein families  \\\n",
       "0                             Cytochrome P450 family   \n",
       "1                                                NaN   \n",
       "2  Protein kinase superfamily, CAMK Ser/Thr prote...   \n",
       "3                                                NaN   \n",
       "4                                       CD300 family   \n",
       "\n",
       "                                   Interacts with  \\\n",
       "0                                             NaN   \n",
       "1                                          Q96S66   \n",
       "2  Q9BRT9; Q68G74; Q14696; P78337; Q9NRG1; P62072   \n",
       "3                                             NaN   \n",
       "4                                             NaN   \n",
       "\n",
       "                                  Compositional bias   3D  \\\n",
       "0                                                NaN  NaN   \n",
       "1                                                NaN  NaN   \n",
       "2  COMPBIAS 457..477; /note=\"Polar residues\"; /ev...  NaN   \n",
       "3                                                NaN  NaN   \n",
       "4                                                NaN  NaN   \n",
       "\n",
       "           Protein existence  \\\n",
       "0                  Uncertain   \n",
       "1  Evidence at protein level   \n",
       "2                  Uncertain   \n",
       "3  Evidence at protein level   \n",
       "4  Evidence at protein level   \n",
       "\n",
       "                                         Domain [FT]  \\\n",
       "0                                                NaN   \n",
       "1                                                NaN   \n",
       "2  DOMAIN 27..278; /note=\"Protein kinase\"; /evide...   \n",
       "3                                                NaN   \n",
       "4  DOMAIN 25..123; /note=\"Ig-like V-type\"; /evide...   \n",
       "\n",
       "           Struct_regions_start..end Struct_regions_fraction  \\\n",
       "0  4..142,148..227,229..329,348..515                    94.8   \n",
       "1                              6..53                    90.7   \n",
       "2                   21..280,303..352                    39.6   \n",
       "3                              6..15                    62.5   \n",
       "4   24..62,81..115,127..139,174..193                    55.2   \n",
       "\n",
       "                     IDR_region_start..end IDR_region_fraction  \n",
       "0      0..3,143..147,228..228,330..347,516                 5.2  \n",
       "1                                  0..5,54                 9.3  \n",
       "2                  0..20,281..302,353..783                60.4  \n",
       "3                                  0..5,16                37.5  \n",
       "4  0..23,63..80,116..126,140..173,194..201                44.8  \n",
       "\n",
       "[5 rows x 28 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "human_df = pd.read_csv('data/Uniprot_raw_human_proteome.tsv', sep='\\t')\n",
    "merged = pd.merge(human_df, mdb, on='Entry',how='outer')\n",
    "merged.dropna(subset = [\"Sequence\"], inplace=True)\n",
    "\n",
    "merged.to_csv('data/Uniprot_human_proteome.tsv.gz', sep='\\t', compression='gzip', index=False)\n",
    "merged.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. Prepare list of human E3 ligases\n",
    "\n",
    "List of human E3 ligases with assigned classes e.g. RING or U-Box was manually prepared based on \n",
    "* [Medvar et al.](https://hpcwebapps.cit.nih.gov/ESBL/Database/E3-ligases/)\n",
    "* [UbiNet 2.0](https://awi.cuhk.edu.cn/~ubinet/index.php)\n",
    "* several other publications\n",
    "\n",
    "It can be found in `data/E3_list.tsv`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4. Prepare list of human housekeeping genes\n",
    "\n",
    "List of human housekeeping genes was obtained from [tau.ac.il](https://www.tau.ac.il/~elieis/HKG/HK_genes.txt)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "mkdir -p data/\n",
    "wget https://www.tau.ac.il/~elieis/HKG/HK_genes.txt -P data/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 5. Predict number of transmembrane helices (TMH)\n",
    "\n",
    "\n",
    "Predict number of transmembrane helices (TMH) for each sequence from all the \"pre-filtered\" MSA files using the standalone version of [TMHMM-2.0](https://services.healthtech.dtu.dk/service.php?TMHMM-2.0) software."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5.1. Prepare fasta file input for the TMHMM-2.0 software\n",
    "\n",
    "Create fasta with all sequences to predict TMH for each of them using the TMHMM-2.0 software.\n",
    "\n",
    "Save all human sequences to one fasta file `files/human_seq_for_TMH.fasta`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('data/Uniprot_human_proteome.tsv.gz', sep='\\t')\n",
    "uniprots = df['Entry'].tolist()\n",
    "sequences = df['Sequence'].tolist()\n",
    "f = open('results/TMH_predictions/human_seq_for_TMH.fasta', 'w')\n",
    "for i in range(len(uniprots)):\n",
    "    f.write(f'>{uniprots[i]}\\n{sequences[i]}\\n')\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###  5.2. Downolad TMHMM-2.0\n",
    "\n",
    "Download the TMHMM-2.0 standalone version [here](https://services.healthtech.dtu.dk/software.php).\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5.3. Run TMHMM-2.0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Provide path to TMHMM\n",
    "\n",
    "Simply write it down in the field below \n",
    "\n",
    "e.g. `/Users/ns/Install/tmhmm-2.0c.Linux/tmhmm-2.0c/bin/tmhmm`\n",
    "\n",
    "**Do not hit ENTER after writing it in the field below.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "db = HBox([Label('Full path to TMHMM-2.0 software:'), Text()])\n",
    "display(db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tmhmm_path = db.children[1].value # get path to TMHMM-2.0 software path from Text object"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Run TMHMM-2.0\n",
    "\n",
    "Run TMHMM-2.0 on `results/TMH_predictions/human_seq_for_TMH.fasta`. Choose short output & save it to `results/TMH_predictions/human_seq_for_TMH.tsv`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$tmhmm_path\"\n",
    "\n",
    "$1 results/TMH_predictions/human_seq_for_TMH.fasta >> results/TMH_predictions/human_seq_for_TMH.tsv\n",
    "gzip -9 results/TMH_predictions/human_seq_for_TMH.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Delete the unnecessary file\n",
    "\n",
    "Delete fasta file with conacatenated human sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "rm results/TMH_predictions/human_seq_for_TMH.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5.4. Assign predicted number of TMH for each human protein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>NumTMH</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A087X1C5</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A0B4J2F0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A0B4J2F2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A0C5B5G6</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A0K2S4Q6</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Entry NumTMH\n",
       "0  A0A087X1C5      2\n",
       "1  A0A0B4J2F0      1\n",
       "2  A0A0B4J2F2      0\n",
       "3  A0A0C5B5G6      0\n",
       "4  A0A0K2S4Q6      1"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmh = pd.read_csv('results/TMH_predictions/human_seq_for_TMH.tsv.gz', sep='\\t', usecols=[0,4], names=['Entry', 'tmp'])\n",
    "\n",
    "tmh['NumTMH'] = [el.split('=')[1] for el in tmh['tmp'].tolist()]\n",
    "tmh = tmh.drop('tmp', axis=1)\n",
    "tmh.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ns/opt/anaconda3/envs/Lys-deserts/lib/python3.9/site-packages/pandas/core/internals/blocks.py:2199: RuntimeWarning: invalid value encountered in cast\n",
      "  values = values.astype(str)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Reviewed</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Protein names</th>\n",
       "      <th>Gene Names</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Length</th>\n",
       "      <th>Proteomes</th>\n",
       "      <th>Subcellular location [CC]</th>\n",
       "      <th>Gene Ontology (biological process)</th>\n",
       "      <th>...</th>\n",
       "      <th>Interacts with</th>\n",
       "      <th>Compositional bias</th>\n",
       "      <th>3D</th>\n",
       "      <th>Protein existence</th>\n",
       "      <th>Domain [FT]</th>\n",
       "      <th>Struct_regions_start..end</th>\n",
       "      <th>Struct_regions_fraction</th>\n",
       "      <th>IDR_region_start..end</th>\n",
       "      <th>IDR_region_fraction</th>\n",
       "      <th>NumTMH</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A087X1C5</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>CP2D7_HUMAN</td>\n",
       "      <td>Putative cytochrome P450 2D7 (EC 1.14.14.1)</td>\n",
       "      <td>CYP2D7</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>515.0</td>\n",
       "      <td>UP000005640: Chromosome 22</td>\n",
       "      <td>SUBCELLULAR LOCATION: Membrane {ECO:0000305}; ...</td>\n",
       "      <td>arachidonic acid metabolic process [GO:0019369...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Uncertain</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4..142,148..227,229..329,348..515</td>\n",
       "      <td>94.8</td>\n",
       "      <td>0..3,143..147,228..228,330..347,516</td>\n",
       "      <td>5.2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A0B4J2F0</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>PIOS1_HUMAN</td>\n",
       "      <td>Protein PIGBOS1 (PIGB opposite strand protein 1)</td>\n",
       "      <td>PIGBOS1</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>54.0</td>\n",
       "      <td>UP000005640: Chromosome 15</td>\n",
       "      <td>SUBCELLULAR LOCATION: Mitochondrion outer memb...</td>\n",
       "      <td>regulation of endoplasmic reticulum unfolded p...</td>\n",
       "      <td>...</td>\n",
       "      <td>Q96S66</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Evidence at protein level</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6..53</td>\n",
       "      <td>90.7</td>\n",
       "      <td>0..5,54</td>\n",
       "      <td>9.3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A0B4J2F2</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>SIK1B_HUMAN</td>\n",
       "      <td>Putative serine/threonine-protein kinase SIK1B...</td>\n",
       "      <td>SIK1B</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>783.0</td>\n",
       "      <td>UP000005640: Chromosome 21</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cellular response to glucose starvation [GO:00...</td>\n",
       "      <td>...</td>\n",
       "      <td>Q9BRT9; Q68G74; Q14696; P78337; Q9NRG1; P62072</td>\n",
       "      <td>COMPBIAS 457..477; /note=\"Polar residues\"; /ev...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Uncertain</td>\n",
       "      <td>DOMAIN 27..278; /note=\"Protein kinase\"; /evide...</td>\n",
       "      <td>21..280,303..352</td>\n",
       "      <td>39.6</td>\n",
       "      <td>0..20,281..302,353..783</td>\n",
       "      <td>60.4</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A0C5B5G6</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>MOTSC_HUMAN</td>\n",
       "      <td>Mitochondrial-derived peptide MOTS-c (Mitochon...</td>\n",
       "      <td>MT-RNR1</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>16.0</td>\n",
       "      <td>UP000005640: Mitochondrion</td>\n",
       "      <td>SUBCELLULAR LOCATION: Secreted {ECO:0000269|Pu...</td>\n",
       "      <td>activation of protein kinase activity [GO:0032...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Evidence at protein level</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6..15</td>\n",
       "      <td>62.5</td>\n",
       "      <td>0..5,16</td>\n",
       "      <td>37.5</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A0K2S4Q6</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>CD3CH_HUMAN</td>\n",
       "      <td>Protein CD300H (CD300 antigen-like family memb...</td>\n",
       "      <td>CD300H</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>201.0</td>\n",
       "      <td>UP000005640: Chromosome 17</td>\n",
       "      <td>SUBCELLULAR LOCATION: [Isoform 1]: Membrane {E...</td>\n",
       "      <td>neutrophil chemotaxis [GO:0030593]</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Evidence at protein level</td>\n",
       "      <td>DOMAIN 25..123; /note=\"Ig-like V-type\"; /evide...</td>\n",
       "      <td>24..62,81..115,127..139,174..193</td>\n",
       "      <td>55.2</td>\n",
       "      <td>0..23,63..80,116..126,140..173,194..201</td>\n",
       "      <td>44.8</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Entry  Reviewed   Entry Name  \\\n",
       "0  A0A087X1C5  reviewed  CP2D7_HUMAN   \n",
       "1  A0A0B4J2F0  reviewed  PIOS1_HUMAN   \n",
       "2  A0A0B4J2F2  reviewed  SIK1B_HUMAN   \n",
       "3  A0A0C5B5G6  reviewed  MOTSC_HUMAN   \n",
       "4  A0A0K2S4Q6  reviewed  CD3CH_HUMAN   \n",
       "\n",
       "                                       Protein names Gene Names  \\\n",
       "0        Putative cytochrome P450 2D7 (EC 1.14.14.1)     CYP2D7   \n",
       "1   Protein PIGBOS1 (PIGB opposite strand protein 1)    PIGBOS1   \n",
       "2  Putative serine/threonine-protein kinase SIK1B...      SIK1B   \n",
       "3  Mitochondrial-derived peptide MOTS-c (Mitochon...    MT-RNR1   \n",
       "4  Protein CD300H (CD300 antigen-like family memb...     CD300H   \n",
       "\n",
       "               Organism  Length                   Proteomes  \\\n",
       "0  Homo sapiens (Human)   515.0  UP000005640: Chromosome 22   \n",
       "1  Homo sapiens (Human)    54.0  UP000005640: Chromosome 15   \n",
       "2  Homo sapiens (Human)   783.0  UP000005640: Chromosome 21   \n",
       "3  Homo sapiens (Human)    16.0  UP000005640: Mitochondrion   \n",
       "4  Homo sapiens (Human)   201.0  UP000005640: Chromosome 17   \n",
       "\n",
       "                           Subcellular location [CC]  \\\n",
       "0  SUBCELLULAR LOCATION: Membrane {ECO:0000305}; ...   \n",
       "1  SUBCELLULAR LOCATION: Mitochondrion outer memb...   \n",
       "2                                                NaN   \n",
       "3  SUBCELLULAR LOCATION: Secreted {ECO:0000269|Pu...   \n",
       "4  SUBCELLULAR LOCATION: [Isoform 1]: Membrane {E...   \n",
       "\n",
       "                  Gene Ontology (biological process)  ...  \\\n",
       "0  arachidonic acid metabolic process [GO:0019369...  ...   \n",
       "1  regulation of endoplasmic reticulum unfolded p...  ...   \n",
       "2  cellular response to glucose starvation [GO:00...  ...   \n",
       "3  activation of protein kinase activity [GO:0032...  ...   \n",
       "4                 neutrophil chemotaxis [GO:0030593]  ...   \n",
       "\n",
       "                                   Interacts with  \\\n",
       "0                                             NaN   \n",
       "1                                          Q96S66   \n",
       "2  Q9BRT9; Q68G74; Q14696; P78337; Q9NRG1; P62072   \n",
       "3                                             NaN   \n",
       "4                                             NaN   \n",
       "\n",
       "                                  Compositional bias   3D  \\\n",
       "0                                                NaN  NaN   \n",
       "1                                                NaN  NaN   \n",
       "2  COMPBIAS 457..477; /note=\"Polar residues\"; /ev...  NaN   \n",
       "3                                                NaN  NaN   \n",
       "4                                                NaN  NaN   \n",
       "\n",
       "           Protein existence  \\\n",
       "0                  Uncertain   \n",
       "1  Evidence at protein level   \n",
       "2                  Uncertain   \n",
       "3  Evidence at protein level   \n",
       "4  Evidence at protein level   \n",
       "\n",
       "                                         Domain [FT]  \\\n",
       "0                                                NaN   \n",
       "1                                                NaN   \n",
       "2  DOMAIN 27..278; /note=\"Protein kinase\"; /evide...   \n",
       "3                                                NaN   \n",
       "4  DOMAIN 25..123; /note=\"Ig-like V-type\"; /evide...   \n",
       "\n",
       "           Struct_regions_start..end Struct_regions_fraction  \\\n",
       "0  4..142,148..227,229..329,348..515                    94.8   \n",
       "1                              6..53                    90.7   \n",
       "2                   21..280,303..352                    39.6   \n",
       "3                              6..15                    62.5   \n",
       "4   24..62,81..115,127..139,174..193                    55.2   \n",
       "\n",
       "                     IDR_region_start..end IDR_region_fraction NumTMH  \n",
       "0      0..3,143..147,228..228,330..347,516                 5.2      2  \n",
       "1                                  0..5,54                 9.3      1  \n",
       "2                  0..20,281..302,353..783                60.4      0  \n",
       "3                                  0..5,16                37.5      0  \n",
       "4  0..23,63..80,116..126,140..173,194..201                44.8      1  \n",
       "\n",
       "[5 rows x 29 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv('data/Uniprot_human_proteome.tsv.gz', sep='\\t')\n",
    "merged = pd.merge(df, tmh, on='Entry',how='outer')\n",
    "merged.to_csv('data/Uniprot_human_proteome.tsv.gz', sep='\\t', compression='gzip', index=False)\n",
    "merged.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 6. Download half-life datasets\n",
    "\n",
    "Download dataset of proteins' half-lives created by Mathieson et al. (2018) - [Supplementary Data 2](https://static-content.springer.com/esm/art%3A10.1038%2Fs41467-018-03106-1/MediaObjects/41467_2018_3106_MOESM5_ESM.xlsx)\n",
    "\n",
    "Download dataset of proteins' half-lives created by Li et al. (2021) - [Supplementary Table 2](https://ars.els-cdn.com/content/image/1-s2.0-S1097276521007498-mmc3.xlsx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "mkdir -p data/half_life\n",
    "wget -O data/half_life/Mathieson_half_life_dataset.xlsx https://static-content.springer.com/esm/art%3A10.1038%2Fs41467-018-03106-1/MediaObjects/41467_2018_3106_MOESM5_ESM.xlsx\n",
    "wget -O data/half_life/Li_half_life_dataset.xlsx https://ars.els-cdn.com/content/image/1-s2.0-S1097276521007498-mmc3.xlsx"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
